

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>limmbo.core.gwas &mdash; LiMMBo b documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="LiMMBo b documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> LiMMBo
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../input.html">Data input</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../vd.html">Variance decomposition</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gwas.html">Association analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cmd.html">Command-line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">Additional functions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">LiMMBo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>limmbo.core.gwas</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for limmbo.core.gwas</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="n">mpl</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s1">&#39;Agg&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">import</span> <span class="nn">copy</span>

<span class="kn">import</span> <span class="nn">scipy</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">import</span> <span class="nn">limix</span> <span class="k">as</span> <span class="nn">limi</span>
<span class="kn">import</span> <span class="nn">limix.qtl</span> <span class="k">as</span> <span class="nn">qtl</span>
<span class="kn">import</span> <span class="nn">limix.plot</span> <span class="k">as</span> <span class="nn">plot</span>
<span class="kn">import</span> <span class="nn">limix.mtset</span>

<span class="kn">from</span> <span class="nn">limmbo.utils.utils</span> <span class="k">import</span> <span class="n">verboseprint</span>
<span class="kn">from</span> <span class="nn">limmbo.utils.utils</span> <span class="k">import</span> <span class="n">effectiveTests</span>

<div class="viewcode-block" id="GWAS"><a class="viewcode-back" href="../../../gwas.html#limmbo.core.gwas.GWAS">[docs]</a><span class="k">class</span> <span class="nc">GWAS</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">datainput</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">searchDelta</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        nothing to initialize</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seed</span> <span class="o">=</span> <span class="n">seed</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">searchDelta</span> <span class="o">=</span> <span class="n">searchDelta</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">genotypes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">genotypes_info</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">phenotypes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">phenotype_ID</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">covariates</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">relatedness</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">pcs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">Cg</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span> <span class="o">=</span> <span class="n">datainput</span><span class="o">.</span><span class="n">Cn</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">test</span> <span class="o">=</span> <span class="s2">&quot;lrt&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">pvalues</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pvalues_adjust</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pvalues_empirical_raw</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stats</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">z</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">model</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pvalues_empirical</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adjustBy</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">estimate_vd</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fdr_empirical</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;datainput.genotypes cannot be coverted to np.array&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;datainput.phenotypes cannot be coverted to np.array&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;datainput.covariates cannot be coverted to np.array&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;datainput.relatedness cannot be coverted to np.array&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="GWAS.runAssociationAnalysis"><a class="viewcode-back" href="../../../gwas.html#limmbo.core.gwas.GWAS.runAssociationAnalysis">[docs]</a>    <span class="k">def</span> <span class="nf">runAssociationAnalysis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">setup</span><span class="o">=</span><span class="s2">&quot;lmm&quot;</span><span class="p">,</span> <span class="n">adjustSingleTrait</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Analysing the association between phenotypes, genotypes, optional</span>
<span class="sd">        covariates and random genetic effects.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            mode (string):</span>
<span class="sd">                pecifies the type of linear model: either &#39;multitrait&#39; for</span>
<span class="sd">                multivariate analysis or &#39;singletrait&#39; for univariate analysis.</span>
<span class="sd">            setup (string):</span>
<span class="sd">                specifies the linear model setup: either &#39;lmm&#39; for linear mixed</span>
<span class="sd">                model or &#39;lm&#39; for a simple linear model.</span>
<span class="sd">            adjustSingleTrait (string):</span>
<span class="sd">                Method to adjust single-trait association p-values for testing</span>
<span class="sd">                multiple traits; If None (default) no adjusting. Options are</span>
<span class="sd">                &#39;bonferroni&#39; (for bonferroni correction&#39;) or &#39;effective&#39; (for</span>
<span class="sd">                correcting for the effective number of tests as described in </span>
<span class="sd">                `(Galwey,2009)</span>
<span class="sd">                &lt;http://onlinelibrary.wiley.com/doi/10.1002/gepi.20408/abstract&gt;`_.</span>
<span class="sd">                </span>

<span class="sd">        Returns:</span>
<span class="sd">            (dictionary):</span>
<span class="sd">                dictionary containing:</span>

<span class="sd">                - **lm** (:class:`limix.qtl.LMM`):</span>
<span class="sd">                  LIMIX LMM object</span>
<span class="sd">                - **pvalues** (numpy array):</span>
<span class="sd">                  [`NrSNP` x `P`] (when mode is singletrait) or [1 x`NrSNP`]</span>
<span class="sd">                  array of p-values.</span>
<span class="sd">                - **betas** (numpy array):</span>
<span class="sd">                  [`NrSNP` x `P`] array of effect size estimates per SNP across</span>
<span class="sd">                  all traits.</span>
<span class="sd">                - **pvalues_adjust** (numpy array):</span>
<span class="sd">                  only returned if mode is &#39;singletrait&#39; and &#39;adjustSingleTrait&#39;</span>
<span class="sd">                  is not None; contains single-trait p-values adjusted for the</span>
<span class="sd">                  number of single-trait analyses conducted.</span>

<span class="sd">        Examples:</span>

<span class="sd">            .. doctest::</span>

<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.reader import ReadData</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.input import InputData</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.core.gwas import GWAS</span>
<span class="sd">                &gt;&gt;&gt; data = ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_pheno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                                &#39;io/test/data/pheno.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; file_geno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                                &#39;io/test/data/genotypes.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; file_relatedness = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/relatedness.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; file_covs = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                               &#39;io/test/data/covs.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; file_Cg = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/Cg.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; file_Cn = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/Cn.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getPhenotypes(file_pheno=file_pheno)</span>
<span class="sd">                &gt;&gt;&gt; data.getCovariates(file_covariates=file_covs)</span>
<span class="sd">                &gt;&gt;&gt; data.getRelatedness(file_relatedness=file_relatedness)</span>
<span class="sd">                &gt;&gt;&gt; data.getGenotypes(file_genotypes=file_geno)</span>
<span class="sd">                &gt;&gt;&gt; data.getVarianceComponents(file_Cg=file_Cg,</span>
<span class="sd">                ...                            file_Cn=file_Cn)</span>
<span class="sd">                &gt;&gt;&gt; indata = InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addPhenotypes(phenotypes = data.phenotypes)</span>
<span class="sd">                &gt;&gt;&gt; indata.addRelatedness(relatedness = data.relatedness)</span>
<span class="sd">                &gt;&gt;&gt; indata.addCovariates(covariates = data.covariates)</span>
<span class="sd">                &gt;&gt;&gt; indata.addGenotypes(genotypes=data.genotypes,</span>
<span class="sd">                ...                     genotypes_info=data.genotypes_info)</span>
<span class="sd">                &gt;&gt;&gt; indata.addVarianceComponents(Cg = data.Cg, Cn=data.Cn)</span>
<span class="sd">                &gt;&gt;&gt; indata.commonSamples()</span>
<span class="sd">                &gt;&gt;&gt; indata.regress()</span>
<span class="sd">                &gt;&gt;&gt; indata.transform(transform=&quot;scale&quot;)</span>
<span class="sd">                &gt;&gt;&gt; gwas = GWAS(datainput=indata, seed=10, verbose=False)</span>
<span class="sd">                &gt;&gt;&gt;</span>
<span class="sd">                &gt;&gt;&gt;</span>
<span class="sd">                &gt;&gt;&gt; # Example of multi-trait single-variant association testing</span>
<span class="sd">                &gt;&gt;&gt; # using a linear mixed model.</span>
<span class="sd">                &gt;&gt;&gt; resultsAssociation = gwas.runAssociationAnalysis(</span>
<span class="sd">                ...     setup=&quot;lmm&quot;, mode=&quot;multitrait&quot;)</span>
<span class="sd">                &gt;&gt;&gt; resultsAssociation.keys()</span>
<span class="sd">                [&#39;lm&#39;, &#39;betas&#39;, &#39;pvalues&#39;]</span>
<span class="sd">                &gt;&gt;&gt; resultsAssociation[&#39;pvalues&#39;].shape</span>
<span class="sd">                (1, 20)</span>
<span class="sd">                &gt;&gt;&gt; &#39;{:0.3e}&#39;.format(resultsAssociation[&#39;pvalues&#39;].min())</span>
<span class="sd">                &#39;4.584e-09&#39;</span>
<span class="sd">                &gt;&gt;&gt; resultsAssociation[&#39;betas&#39;].shape</span>
<span class="sd">                (10, 20)</span>
<span class="sd">                &gt;&gt;&gt;</span>
<span class="sd">                &gt;&gt;&gt;</span>
<span class="sd">                &gt;&gt;&gt; # Example of single-trait single-variant association testing</span>
<span class="sd">                &gt;&gt;&gt; # using a linear mixed model.</span>
<span class="sd">                &gt;&gt;&gt; resultsAssociation = gwas.runAssociationAnalysis(</span>
<span class="sd">                ...     setup=&quot;lmm&quot;, mode=&quot;singletrait&quot;,</span>
<span class="sd">                ...     adjustSingleTrait = &quot;effective&quot;)</span>
<span class="sd">                &gt;&gt;&gt; resultsAssociation.keys()</span>
<span class="sd">                [&#39;pvalues_adjust&#39;, &#39;lm&#39;, &#39;betas&#39;, &#39;pvalues&#39;]</span>
<span class="sd">                &gt;&gt;&gt; resultsAssociation[&#39;pvalues&#39;].shape</span>
<span class="sd">                (10, 20)</span>
<span class="sd">                &gt;&gt;&gt; resultsAssociation[&#39;betas&#39;].shape</span>
<span class="sd">                (10, 20)</span>
<span class="sd">                &gt;&gt;&gt; &#39;{:0.3e}&#39;.format(resultsAssociation[&#39;pvalues_adjust&#39;].min())</span>
<span class="sd">                &#39;2.262e-03&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>


        <span class="c1"># set parameters for the analysis</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">P</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">S</span>    <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Loaded </span><span class="si">{}</span><span class="s2"> samples, </span><span class="si">{}</span><span class="s2"> phenotypes, </span><span class="si">{}</span><span class="s2"> snps&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">P</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">S</span><span class="p">),</span> <span class="n">verbose</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            
        <span class="bp">self</span><span class="o">.</span><span class="n">setup</span> <span class="o">=</span> <span class="n">setup</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="n">mode</span>

        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;multitrait&quot;</span><span class="p">:</span>
            <span class="n">associationResults</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__multiTraitAssociation_anyeffect</span><span class="p">(</span>
                <span class="n">genotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;singletrait&quot;</span><span class="p">:</span>
            <span class="n">associationResults</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__singleTraitAssociation</span><span class="p">(</span>
                <span class="n">genotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="p">,</span>
                <span class="n">adjustSingleTrait</span> <span class="o">=</span> <span class="n">adjustSingleTrait</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">associationResults</span></div>

    <span class="k">def</span> <span class="nf">__multiTraitAssociation_anyeffect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">,</span> <span class="n">empiricalP</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">computeFDR</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Muti-trait association test, testing for an effect of the genotype on</span>
<span class="sd">        `any` phenotype (`any` effect test). Wraps around</span>
<span class="sd">        `qtl.qtl_test_lmm_kronecker`.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            genotypes (array-like):</span>
<span class="sd">                [`N` x `NrSNP`] genotype array of [`N`] samples and [`NrSNP`]</span>
<span class="sd">                genotypes</span>
<span class="sd">            empiricalP (bool):</span>
<span class="sd">                set to True if association test is part of estimating empirical</span>
<span class="sd">                pvalues</span>
<span class="sd">            computeFDR (bool):</span>
<span class="sd">                set to True if association test is part of estimating empirical</span>
<span class="sd">                FDR</span>

<span class="sd">        Returns:</span>
<span class="sd">            (dictionary):</span>
<span class="sd">                dictionary containing:</span>

<span class="sd">                - **lm**(:class:`limix.qtl.LMM`):</span>
<span class="sd">                  LIMIX LMM object</span>
<span class="sd">                - **pvalues** (numpy array):</span>
<span class="sd">                  [1 x`NrSNP`] array of</span>
<span class="sd">                  p-values</span>
<span class="sd">                - **betas** (numpy array):</span>
<span class="sd">                  [`NrSNP` x `P`] array of effect size estimates per SNP across</span>
<span class="sd">                  all traits</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">Acovs</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">Acovs</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">setup</span> <span class="ow">is</span> <span class="s2">&quot;lmm&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">P</span> <span class="o">&gt;</span> <span class="mi">30</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: For large trait sizes, computation times &quot;</span>
                  <span class="s2">&quot;for pure REML variance decomposition are long, &quot;</span>
                   <span class="s2">&quot;consider bootstrapping trait-trait covariance&quot;</span>
                    <span class="s2">&quot; components&quot;</span><span class="p">)</span>
                <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Estimate Variance components&quot;</span><span class="p">,</span>
                            <span class="n">verbose</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">__varianceDecomposition</span><span class="p">()</span>
            <span class="n">K1c</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span>
            <span class="n">K2c</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span>
            <span class="n">K1r</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">=</span><span class="s2">&quot;lmm_mt&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">K1c</span> <span class="o">=</span> <span class="mf">1e-9</span><span class="o">*</span><span class="n">sp</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P</span><span class="p">)</span>
            <span class="n">K2c</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">cov</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
            <span class="n">K1r</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">=</span><span class="s2">&quot;lm_mt_pcs&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">=</span><span class="s2">&quot;lm_mt&quot;</span>
        <span class="n">Asnps</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">empiricalP</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">computeFDR</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Computing multi-trait (any effect) model: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">),</span> <span class="n">verbose</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>

        <span class="n">lm</span><span class="p">,</span> <span class="n">pvalues</span> <span class="o">=</span> <span class="n">qtl</span><span class="o">.</span><span class="n">qtl_test_lmm_kronecker</span><span class="p">(</span><span class="n">snps</span><span class="o">=</span><span class="n">genotypes</span><span class="p">,</span>
                <span class="n">phenos</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">,</span> <span class="n">Asnps</span><span class="o">=</span><span class="n">Asnps</span><span class="p">,</span> <span class="n">Acovs</span><span class="o">=</span><span class="n">Acovs</span><span class="p">,</span>
                <span class="n">covs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="p">,</span> <span class="n">K1r</span><span class="o">=</span><span class="n">K1r</span><span class="p">,</span> <span class="n">K1c</span><span class="o">=</span><span class="n">K1c</span><span class="p">,</span> <span class="n">K2c</span><span class="o">=</span><span class="n">K2c</span><span class="p">,</span>
                <span class="n">searchDelta</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">searchDelta</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">empiricalP</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">computeFDR</span><span class="p">:</span>
            <span class="n">betas</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">getBetaSNP</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">betas</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">lm</span> <span class="o">=</span> <span class="kc">None</span>
        
        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;pvalues&quot;</span><span class="p">:</span> <span class="n">pvalues</span><span class="p">,</span> <span class="s2">&quot;betas&quot;</span><span class="p">:</span> <span class="n">betas</span><span class="p">,</span> <span class="s2">&quot;lm&quot;</span><span class="p">:</span> <span class="n">lm</span><span class="p">}</span>


    <span class="k">def</span> <span class="nf">__singleTraitAssociation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">,</span> <span class="n">adjustSingleTrait</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">empiricalP</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">computeFDR</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Single-trait association test. Wraps around `qtl.qtl_test_lmm`.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            genotypes (array-like):</span>
<span class="sd">                [`N` x `NrSNP`] genotype array of [`N`] samples and [`NrSNP`]</span>
<span class="sd">                genotypes</span>
<span class="sd">            empiricalP (bool):</span>
<span class="sd">                set to True if association test is part of estimating empirical</span>
<span class="sd">                pvalues</span>
<span class="sd">            computeFDR (bool):</span>
<span class="sd">                set to True if association test is part of estimating empirical</span>
<span class="sd">                FDR</span>
<span class="sd">            adjustSingleTrait (string):</span>
<span class="sd">                Method to adjust single-trait association p-values for testing</span>
<span class="sd">                multiple traits; If None (default) no adjusting. Options are</span>
<span class="sd">                &#39;bonferroni&#39; (for bonferroni correction&#39;) or &#39;effective&#39; (for</span>
<span class="sd">                correcting for the effective number of tests as described in </span>
<span class="sd">                `(Galwey,2009) &lt;http://onlinelibrary.wiley.com/doi/10.1002/gepi.20408/abstract&gt;`_.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (dictionary):</span>
<span class="sd">                dictionary containing:</span>

<span class="sd">                    - **lm**(:class:`limix.qtl.LMM`):</span>
<span class="sd">                      LIMIX LMM object</span>
<span class="sd">                    - **pvalues** (numpy array):</span>
<span class="sd">                      [`P` x `NrSNP`] array of p-values</span>
<span class="sd">                    - **betas** (numpy array):</span>
<span class="sd">                      [`P` x `NrSNP`] array of effect size estimates per SNP </span>
<span class="sd">                      across all traits</span>
<span class="sd">                    - **pvalues_adjust** (numpy array):</span>
<span class="sd">                      only returned if &#39;adjustSingleTrait&#39; is not None; contains</span>
<span class="sd">                      single-trait p-values adjusted for the number of </span>
<span class="sd">                      single-trait analyses conducted</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">setup</span> <span class="ow">is</span> <span class="s2">&quot;lmm&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">=</span><span class="s2">&quot;lmm_st&quot;</span>
            <span class="n">K</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">=</span><span class="s2">&quot;lm_st_pcs&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">=</span><span class="s2">&quot;lm_st&quot;</span>
                <span class="n">K</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">empiricalP</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">computeFDR</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Computing single-trait association (</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">),</span> <span class="n">verbose</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="n">lm</span> <span class="o">=</span> <span class="n">qtl</span><span class="o">.</span><span class="n">qtl_test_lmm</span><span class="p">(</span><span class="n">snps</span><span class="o">=</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">pheno</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">,</span>
                <span class="n">K</span><span class="o">=</span><span class="n">K</span><span class="p">,</span> <span class="n">covs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">test</span><span class="p">)</span>
            <span class="n">pvalues</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">getPv</span><span class="p">()</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">empiricalP</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">computeFDR</span><span class="p">:</span>
            <span class="n">betas</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">getBetaSNP</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">betas</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">lm</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">adjustSingleTrait</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">adjustBy</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">adjustSingleTrait</span> <span class="ow">is</span> <span class="s2">&quot;bonferroni&quot;</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">adjustBy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P</span>
                <span class="k">elif</span> <span class="n">adjustSingleTrait</span> <span class="ow">is</span> <span class="s2">&quot;effective&quot;</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">adjustBy</span> <span class="o">=</span> <span class="n">effectiveTests</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="p">((</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> is not a provided method to adjust single-trait&quot;</span>
                    <span class="s2">&quot;pvalues for multiple hypothesis testing&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">adjustSingleTrait</span><span class="p">))</span>
            <span class="n">pvalues_adjust</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">__adjust</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pvalues</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">pvalues_adjust</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">adjustSingleTrait</span> <span class="o">=</span> <span class="n">adjustSingleTrait</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;lm&quot;</span><span class="p">:</span><span class="n">lm</span><span class="p">,</span> <span class="s2">&quot;pvalues&quot;</span><span class="p">:</span> <span class="n">pvalues</span><span class="p">,</span> <span class="s2">&quot;pvalues_adjust&quot;</span><span class="p">:</span> <span class="n">pvalues_adjust</span><span class="p">,</span>
                <span class="s2">&quot;betas&quot;</span><span class="p">:</span> <span class="n">betas</span><span class="p">}</span>


<div class="viewcode-block" id="GWAS.saveAssociationResults"><a class="viewcode-back" href="../../../gwas.html#limmbo.core.gwas.GWAS.saveAssociationResults">[docs]</a>    <span class="k">def</span> <span class="nf">saveAssociationResults</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">results</span><span class="p">,</span> <span class="n">outdir</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="n">pvalues_empirical</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Saves results of association analyses.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            results (dictionary):</span>
<span class="sd">                dictionary generated via runAssociation analysis, containing:</span>
<span class="sd">		</span>
<span class="sd">                - **lm** (:class:`limix.qtl.LMM`):</span>
<span class="sd">		  LIMIX LMM object</span>
<span class="sd">		- **pvalues** (numpy array):</span>
<span class="sd">		  [`P` x `NrSNP`] array of p-values</span>
<span class="sd">		- **betas** (numpy array):</span>
<span class="sd">		  [`P` x `NrSNP`] array of effect size estimates per SNP</span>
<span class="sd">		  across all traits</span>
<span class="sd">		- **pvalues_adjust** (numpy array):</span>
<span class="sd">		  only returned mode == singletrait and if &#39;adjustSingleTrait&#39;</span>
<span class="sd">                  is not None; contains single-trait p-values adjusted for the </span>
<span class="sd">                  number of single-trait analyses conducted</span>
<span class="sd">                </span>
<span class="sd">            outdir (string):</span>
<span class="sd">                &#39;/path/to/output/directory&#39;; needs user writing permission</span>
<span class="sd">            name (string):</span>
<span class="sd">                input name specific name, such as chromosome used in analysis</span>
<span class="sd">            pvalues_empirical (pandas dataframe, optional):</span>
<span class="sd">                empirical pvalues via adjustSingleTrait </span>

<span class="sd">        Returns:</span>
<span class="sd">            (None)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
            <span class="n">outstring</span> <span class="o">=</span> <span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">name</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">outstring</span> <span class="o">=</span> <span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span>

        <span class="n">beta_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;betas&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">)</span>
        <span class="n">beta_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">,</span> <span class="n">beta_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="ow">is</span> <span class="s1">&#39;singletrait&#39;</span><span class="p">:</span>
            <span class="n">pvalues_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>  
                <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">)</span>
            <span class="n">pvalues_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">,</span> <span class="n">pvalues_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues_adjust&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">pvalues_adjust_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues_adjust&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>  
                    <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> 
                    <span class="n">columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">)</span>
                <span class="n">pvalues_adjust_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">,</span>
                    <span class="n">pvalues_adjust_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                
                <span class="n">pvalues_adjust_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">_padjust</span><span class="si">%s</span><span class="s2">.csv&quot;</span> <span class="o">%</span> <span class="n">outstring</span><span class="p">,</span>
                    <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            
            <span class="k">if</span> <span class="n">pvalues_empirical</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                
                <span class="n">pempirical_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pvalues_empirical</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>
                    <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> 
                    <span class="n">columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">)</span>
                    <span class="c1">#index=SNP, columns=[&#39;Pempirical&#39;])</span>
                <span class="n">pempirical_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">,</span> 
                    <span class="n">pempirical_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                
                <span class="n">pempirical_raw_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pvalues_empirical_raw</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>
                    <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> 
                    <span class="n">columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">)</span>
                <span class="n">pempirical_raw_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">,</span> 
                    <span class="n">pempirical_raw_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                
                <span class="n">pempirical_raw_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">_pempirical_raw</span><span class="si">%s%s</span><span class="s2">.csv&quot;</span> <span class="o">%</span>
                    <span class="p">(</span><span class="n">outstring</span> <span class="o">+</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fdr</span><span class="p">,)),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="ow">is</span> <span class="s1">&#39;multitrait&#39;</span><span class="p">:</span>
            <span class="n">pvalues_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>
                <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;P&quot;</span><span class="p">])</span>
            <span class="n">pvalues_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">,</span> <span class="n">pvalues_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            
            <span class="k">if</span> <span class="n">pvalues_empirical</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">pvalues_empirical_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pvalues_empirical</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>
                    <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;P&quot;</span><span class="p">])</span>
                <span class="n">pvalues_empirical_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">,</span> 
                    <span class="n">pvalues_empirical</span><span class="o">-</span><span class="n">df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

            
        <span class="n">pvalues_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">_pvalue</span><span class="si">%s</span><span class="s2">.csv&quot;</span> <span class="o">%</span> <span class="n">outstring</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">beta_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">_betavalue</span><span class="si">%s</span><span class="s2">.csv&quot;</span> <span class="o">%</span> <span class="n">outstring</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">pvalues_empirical</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">pempirical_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">_pempirical</span><span class="si">%s%s</span><span class="s2">.csv&quot;</span> <span class="o">%</span>
                <span class="p">(</span><span class="n">outstring</span> <span class="o">+</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fdr</span><span class="p">,)),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">estimate_vd</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">timeVD</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">timeVD</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                    <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/timeVarianceDecomposition_REML.csv&quot;</span> <span class="o">%</span> <span class="n">outdir</span><span class="p">,</span>
                    <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/Cg_REML.csv&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">outdir</span><span class="p">),</span>
                <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/Cn_REML.csv&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">outdir</span><span class="p">),</span>
                <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">fdr</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppermute</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">_ppermute</span><span class="si">%s</span><span class="s2">.csv&quot;</span> <span class="o">%</span>
                    <span class="n">outstring</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="s1">&#39;FDR&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">FDR</span><span class="p">)])</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                    <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">_empiricalFDR_</span><span class="si">%s</span><span class="s2">.csv&quot;</span> <span class="o">%</span> <span class="n">outstring</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
                    <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></div>

<div class="viewcode-block" id="GWAS.computeEmpiricalP"><a class="viewcode-back" href="../../../gwas.html#limmbo.core.gwas.GWAS.computeEmpiricalP">[docs]</a>    <span class="k">def</span> <span class="nf">computeEmpiricalP</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pvalues</span><span class="p">,</span> <span class="n">nrpermutations</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Compute empirical p-values: permute the genotypes, do the </span>
<span class="sd">        association test, record if permuted p-value of SNP is smaller than</span>
<span class="sd">        original p-value. Sum these occurrences and divide by total number of</span>
<span class="sd">        permutation.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            pvalues (array-like):</span>
<span class="sd">                [`P` x `NrSNP`] (single-trait) or [1 x `NrSNP`] (multi-trait) </span>
<span class="sd">                array of p-values</span>
<span class="sd">            nrpermutations (int):</span>
<span class="sd">                number of permutations; 1/nrpermutations is the maximum level</span>
<span class="sd">                of significance (alpha)to test for, e.g. </span>
<span class="sd">                nrpermuations=100 -&gt; alpha=0.01 </span>

<span class="sd">        Returns:</span>
<span class="sd">            (numpy array): </span>
<span class="sd">                [`P` x `NrSNP`] (single-trait) or [1 x `NrSNP`] (multi-trait) </span>
<span class="sd">                array of empirical p-values</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Computing empirical p-values&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
       
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">seed</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">nrpermutations</span> <span class="o">=</span> <span class="n">nrpermutations</span>
        <span class="n">pvalues</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pvalues</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="ow">is</span> <span class="s2">&quot;multitrait&quot;</span><span class="p">:</span>
            <span class="n">count</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">nrpermutations</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">S</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">count</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">nrpermutations</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">P</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">S</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">ps</span> <span class="ow">in</span> <span class="nb">range</span> <span class="p">(</span><span class="n">nrpermutations</span><span class="p">):</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Permutation </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">ps</span><span class="p">)</span>
            <span class="n">genotypes_permute</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">,</span> 
                <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="p">:]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="ow">is</span> <span class="s2">&quot;multitrait&quot;</span><span class="p">:</span>
                <span class="n">resultsPermute</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__multiTraitAssociation_anyeffect</span><span class="p">(</span> 
                    <span class="n">genotypes</span> <span class="o">=</span> <span class="n">genotypes_permute</span><span class="p">,</span> <span class="n">empiricalP</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">count</span><span class="p">[</span><span class="n">ps</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">resultsPermute</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">pvalues</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="ow">is</span> <span class="s2">&quot;singletrait&quot;</span><span class="p">:</span>
                <span class="n">resultsPermute</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__singletraitAssociation</span><span class="p">(</span>
                    <span class="n">genotypes</span> <span class="o">=</span> <span class="n">genotypes_permute</span><span class="p">,</span> <span class="n">empiricalP</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">count</span><span class="p">[</span><span class="n">ps</span><span class="p">,:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">resultsPermute</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">pvalues</span>


        <span class="n">pvalues_empirical</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">/</span><span class="n">nrpermutations</span>
        
        <span class="k">return</span> <span class="n">pvalues_empirical</span></div>

<div class="viewcode-block" id="GWAS.computeFDR"><a class="viewcode-back" href="../../../gwas.html#limmbo.core.gwas.GWAS.computeFDR">[docs]</a>    <span class="k">def</span> <span class="nf">computeFDR</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create an empirical p-values distribution by permuting the genotypes</span>
<span class="sd">        and running the association test on these (10 random permuations). </span>
<span class="sd">        Recored the observed pvalues, order by rank and find the rank of the </span>
<span class="sd">        desired FDR to determine the empirical FDR.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            fdr (float):</span>
<span class="sd">                desired fdr threshold</span>

<span class="sd">        Returns:</span>
<span class="sd">            (tuple):</span>
<span class="sd">                tuple containing:</span>

<span class="sd">                    - **empirical fdr** (float):</span>
<span class="sd">                    - **empirical_pvalue_distribution** (numpy array):</span>
<span class="sd">                      array of empirical p-values for 10 permutations tests</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Computing empirical FDR&quot;</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
        
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">seed</span><span class="p">)</span>
        <span class="c1">#tests = int(1/self.options.fdr)</span>
        <span class="n">tests</span> <span class="o">=</span> <span class="mi">10</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ppermute</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">tests</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">S</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">ps</span> <span class="ow">in</span> <span class="nb">range</span> <span class="p">(</span><span class="n">tests</span><span class="p">):</span>
            <span class="n">genotypes_permute</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">,</span>
                <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="p">:]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;multitrait&quot;</span><span class="p">:</span>
                <span class="n">resultsFDR</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__multiTraitAssociation_anyeffect</span><span class="p">(</span>
                        <span class="n">genotypes</span> <span class="o">=</span> <span class="n">genotypes_permute</span><span class="p">,</span> <span class="n">computeFDR</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">ppermute</span><span class="p">[</span><span class="n">ps</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">resultsFDR</span><span class="p">[</span><span class="s1">&#39;pvalues_permute&#39;</span><span class="p">]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;singletrait&quot;</span><span class="p">:</span>
                <span class="n">resultsFDR</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__singleTraitAssociation</span><span class="p">(</span>
                    <span class="n">genotypes</span> <span class="o">=</span> <span class="n">genotypes_permute</span><span class="p">,</span> <span class="n">computeFDR</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">adjustSingleTrait</span><span class="o">=</span><span class="n">adjustSingleTrait</span><span class="p">)</span>
                <span class="n">pvalues_adjust_min</span> <span class="o">=</span> <span class="n">resultsFDR</span><span class="p">[</span><span class="s1">&#39;pvalues_permute_adjust&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(</span>
                    <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="n">ppermute</span><span class="p">[</span><span class="n">ps</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">pvalues_adjust_min</span>

        <span class="n">SNPsPassingFDR</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fdr</span><span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">S</span><span class="o">*</span> <span class="n">tests</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">allppermute</span> <span class="o">=</span> <span class="n">ppermute</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">allppermute</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fdr_empirical</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">allppermute</span><span class="p">[</span><span class="n">SNPsPassingFDR</span><span class="p">]</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">fdr_empirical</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">allppermute</span></div>

    <span class="k">def</span> <span class="nf">__adjust</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pv</span><span class="p">):</span>
	<span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Adjust pvalues for multiple hypothesis testing by multiplying with </span>
<span class="sd">        a constant (if adjustBy=P, equivalent to Bonferroni adjustment).</span>

<span class="sd">	Arguments:</span>
<span class="sd">            pv (array like):</span>
<span class="sd">                p-values</span>

<span class="sd">        Returns:</span>
<span class="sd">            pv (numpy array):</span>
<span class="sd">                adjusted p-values</span>
<span class="sd">	&quot;&quot;&quot;</span>
        <span class="n">pvadjust</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">min</span><span class="p">(</span><span class="n">pveff</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">pveff</span> <span class="ow">in</span> <span class="p">(</span><span class="n">pv</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">adjustBy</span><span class="p">)])</span>
        <span class="k">return</span> <span class="n">pvadjust</span>

    <span class="k">def</span> <span class="nf">__varianceDecomposition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cache</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="n">vd</span> <span class="o">=</span> <span class="n">limix</span><span class="o">.</span><span class="n">mtset</span><span class="o">.</span><span class="n">MTSet</span><span class="p">(</span><span class="n">Y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">,</span> <span class="n">R</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span><span class="p">)</span>
        <span class="n">vd_null_info</span> <span class="o">=</span> <span class="n">vd</span><span class="o">.</span><span class="n">fitNull</span><span class="p">(</span><span class="n">n_times</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">rewrite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">vd_null_info</span><span class="p">[</span><span class="s1">&#39;conv&#39;</span><span class="p">]:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Variance decomposition converged&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span> <span class="o">=</span> <span class="n">vd_null_info</span><span class="p">[</span><span class="s1">&#39;Cg&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="mf">1e-4</span><span class="o">*</span><span class="n">sp</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span> <span class="o">=</span> <span class="n">vd_null_info</span><span class="p">[</span><span class="s1">&#39;Cn&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="mf">1e-4</span><span class="o">*</span><span class="n">sp</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">time</span> <span class="o">=</span> <span class="n">vd_null_info</span><span class="p">[</span><span class="s1">&#39;time&#39;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s2">&quot;Variance decomposition did not converge&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="GWAS.manhattanQQ"><a class="viewcode-back" href="../../../gwas.html#limmbo.core.gwas.GWAS.manhattanQQ">[docs]</a>    <span class="k">def</span> <span class="nf">manhattanQQ</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">results</span><span class="p">,</span> <span class="n">colourS</span><span class="o">=</span><span class="s1">&#39;DarkBLue&#39;</span><span class="p">,</span> <span class="n">colourNS</span><span class="o">=</span><span class="s1">&#39;Orange&#39;</span><span class="p">,</span>
            <span class="n">alphaS</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">alphaNS</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">thr_plotting</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">saveTo</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Plot manhattan and quantile-quantile plot of association results.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            results (dictionary):</span>
<span class="sd">                dictionary generated via runAssociation analysis</span>
<span class="sd">            colourS (string, optional):</span>
<span class="sd">                colour of significant points</span>
<span class="sd">            colourNS (string, optional):</span>
<span class="sd">                colour of non-significant points</span>
<span class="sd">            alphaS (float, optional):</span>
<span class="sd">                plotting transparency of significant points</span>
<span class="sd">            alphaNS (float, optional):</span>
<span class="sd">                plotting transparency of non-significant points</span>
<span class="sd">            thr_plotting (float, optional):</span>
<span class="sd">                y-intercept for horizontal line as a marker for significance</span>
<span class="sd">            saveTo (string, optional):</span>
<span class="sd">		/path/to/output/directory to automatically save plot as pdf;</span>
<span class="sd">                needs user writing permission.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (None)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="ow">is</span> <span class="s1">&#39;singletrait&#39;</span><span class="p">:</span>
            <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">12</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
            <span class="n">pv_min</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;pv&#39;</span><span class="p">:</span> <span class="n">pv_min</span><span class="p">,</span> 
                <span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">),</span> 
                <span class="s1">&#39;pos&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">[</span><span class="s1">&#39;pos&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)})</span>
            <span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">plot</span><span class="o">.</span><span class="n">plot_manhattan</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> 
                <span class="n">null_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;alpha&#39;</span><span class="p">:</span><span class="n">alphaS</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">colourS</span><span class="p">},</span> 
                <span class="n">alt_style</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;alpha&#39;</span><span class="p">:</span><span class="n">alphaNS</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">colourNS</span><span class="p">},</span>
                <span class="n">alpha</span><span class="o">=</span><span class="n">thr_plotting</span><span class="p">)</span>
            <span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
            <span class="n">plot</span><span class="o">.</span><span class="n">qqplot</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="n">alphaLevel</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>      
            
            <span class="k">if</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues_adjust&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            	<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">12</span><span class="p">,</span><span class="mi">8</span><span class="p">])</span>
            	<span class="n">pv_adjust_min</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues_adjust&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">min</span><span class="p">(</span>
                    <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
                <span class="n">data_adjust</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;pv&#39;</span><span class="p">:</span> <span class="n">pv_min</span><span class="p">,</span> 
                    <span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">),</span> 
                    <span class="s1">&#39;pos&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">[</span><span class="s1">&#39;pos&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)})</span>
		<span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">plot</span><span class="o">.</span><span class="n">plot_manhattan</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> 
                    <span class="n">null_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;alpha&#39;</span><span class="p">:</span><span class="n">alphaS</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">colourS</span><span class="p">},</span> 
                    <span class="n">alt_style</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;alpha&#39;</span><span class="p">:</span><span class="n">alphaNS</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">colourNS</span><span class="p">},</span>
                    <span class="n">alpha</span><span class="o">=</span><span class="n">thr_plotting</span><span class="p">)</span>
		<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> (p-values)&#39;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
		<span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
		<span class="n">plot</span><span class="o">.</span><span class="n">qqplot</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">ravel</span><span class="p">())</span>
		<span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
                <span class="n">plot</span><span class="o">.</span><span class="n">plot_manhattan</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">data_adjust</span><span class="p">,</span> 
                    <span class="n">null_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;alpha&#39;</span><span class="p">:</span><span class="n">alphaS</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">colourS</span><span class="p">},</span> 
                    <span class="n">alt_style</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;alpha&#39;</span><span class="p">:</span><span class="n">alphaNS</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">colourNS</span><span class="p">},</span>
                    <span class="n">alpha</span><span class="o">=</span><span class="n">thr_plotting</span><span class="p">)</span>
		<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> (p-values adjust)&#39;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
		<span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
		<span class="n">plot</span><span class="o">.</span><span class="n">qqplot</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues_adjust&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">ravel</span><span class="p">())</span>
		<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>      
  
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="ow">is</span> <span class="s1">&#39;multitrait&#39;</span><span class="p">:</span>
            <span class="n">pv</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span> 
            <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;pv&#39;</span><span class="p">:</span> <span class="n">pv</span><span class="p">,</span> 
                <span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">),</span> 
                <span class="s1">&#39;pos&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">[</span><span class="s1">&#39;pos&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)})</span>
            <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">12</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
            <span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">plot</span><span class="o">.</span><span class="n">plot_manhattan</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> 
                <span class="n">null_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;alpha&#39;</span><span class="p">:</span><span class="n">alphaS</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">colourS</span><span class="p">},</span> 
                <span class="n">alt_style</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;alpha&#39;</span><span class="p">:</span><span class="n">alphaNS</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">colourNS</span><span class="p">},</span>
                <span class="n">alpha</span><span class="o">=</span><span class="n">thr_plotting</span><span class="p">)</span>
            <span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
            <span class="n">plot</span><span class="o">.</span><span class="n">qqplot</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">ravel</span><span class="p">())</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">saveTo</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">output</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.png&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">saveTo</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">output</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">.png&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">saveTo</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output</span><span class="p">)</span></div></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Hannah V. Meyer.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'b',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>