

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>limmbo.io.input &mdash; LiMMBo b documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="LiMMBo b documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> LiMMBo
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../input.html">Data input</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../vd.html">Variance decomposition</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gwas.html">Association analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cmd.html">Command-line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">Additional functions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">LiMMBo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>limmbo.io.input</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for limmbo.io.input</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">limix.util.preprocess</span> <span class="k">import</span> <span class="n">regressOut</span>

<span class="kn">from</span> <span class="nn">limmbo.utils.utils</span> <span class="k">import</span> <span class="n">verboseprint</span>
<span class="kn">from</span> <span class="nn">limmbo.utils.utils</span> <span class="k">import</span> <span class="n">match</span>
<span class="kn">from</span> <span class="nn">limmbo.utils.utils</span> <span class="k">import</span> <span class="n">scale</span>
<span class="kn">from</span> <span class="nn">limmbo.utils.utils</span> <span class="k">import</span> <span class="n">makeHardCalledGenotypes</span>
<span class="kn">from</span> <span class="nn">limmbo.utils.utils</span> <span class="k">import</span> <span class="n">AlleleFrequencies</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">re</span>

<span class="kn">from</span> <span class="nn">scipy_sugar.stats</span> <span class="k">import</span> <span class="n">quantile_gaussianize</span>
<span class="kn">from</span> <span class="nn">math</span> <span class="k">import</span> <span class="n">sqrt</span>

<span class="k">class</span> <span class="nc">MissingInput</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Raised when no appropriate input is given&quot;&quot;&quot;</span>
    <span class="k">pass</span>


<span class="k">class</span> <span class="nc">FormatError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Raised when no appropriate input is given&quot;&quot;&quot;</span>
    <span class="k">pass</span>


<span class="k">class</span> <span class="nc">DataMismatch</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Raised when dimensions of sample/ID names do not match dimension of</span>
<span class="sd">    corresponding data&quot;&quot;&quot;</span>
    <span class="k">pass</span>


<div class="viewcode-block" id="InputData"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData">[docs]</a><span class="k">class</span> <span class="nc">InputData</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate object containing all datasets relevant for variance decomposition</span>
<span class="sd">    (phenotypes, relatedness estimates) and pre-processing steps (check for</span>
<span class="sd">    common samples and sample order, covariates regression and phenotype</span>
<span class="sd">    transformation)</span>

<span class="sd">    Arguments:</span>
<span class="sd">        verbose (bool):</span>
<span class="sd">            initialise verbose: should progress messages be printed to stdout</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">covariate_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">snps</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="InputData.addPhenotypes"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.addPhenotypes">[docs]</a>    <span class="k">def</span> <span class="nf">addPhenotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenotypes</span><span class="p">,</span> <span class="n">pheno_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">phenotype_ID</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add phenotypes, their phenotype ID and their sample IDs to</span>
<span class="sd">        InputData instance</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            phenotypes (array-like):</span>
<span class="sd">                [`N x `P`] phenotype matrix of `N` individuals and `P`</span>
<span class="sd">                phenotypes; if pandas.DataFrame with pheno_samples as index and </span>
<span class="sd">                phenotypes_ID as columns, pheno_samples and phenotype_ID do not </span>
<span class="sd">                have to specified separately.</span>
<span class="sd">            pheno_samples (array-like):</span>
<span class="sd">                [`N`] sample ID</span>
<span class="sd">            phenotype_ID (array-like):</span>
<span class="sd">                [`P`] phenotype IDs</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.phenotypes** (pd.DataFrame):</span>
<span class="sd">                  [`N` x `P`] phenotype array</span>
<span class="sd">                - **self.pheno_samples** (np.array):</span>
<span class="sd">                  [`N`] sample IDs</span>
<span class="sd">                - **self.phenotype_ID** (np.array):</span>
<span class="sd">                  [`P`] phenotype IDs</span>
<span class="sd">        </span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; import numpy as np</span>
<span class="sd">                &gt;&gt;&gt; import pandas as pd</span>
<span class="sd">                &gt;&gt;&gt; pheno = np.array(((1,2),(7,1),(3,4)))</span>
<span class="sd">                &gt;&gt;&gt; pheno_samples = [&#39;S1&#39;,&#39;S2&#39;, &#39;S3&#39;]</span>
<span class="sd">                &gt;&gt;&gt; phenotype_ID = [&#39;ID1&#39;,&#39;ID2&#39;]</span>
<span class="sd">                &gt;&gt;&gt; phenotypes = pd.DataFrame(pheno, index=pheno_samples,</span>
<span class="sd">                ...     columns = phenotype_ID)</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addPhenotypes(phenotypes = phenotypes)</span>
<span class="sd">                &gt;&gt;&gt; print indata.phenotypes.shape</span>
<span class="sd">                (3, 2)</span>
<span class="sd">                &gt;&gt;&gt; print indata.pheno_samples.shape</span>
<span class="sd">                (3,)</span>
<span class="sd">                &gt;&gt;&gt; print indata.phenotype_ID.shape</span>
<span class="sd">                (2,)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">pheno_samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">((</span><span class="s2">&quot;pheno_samples are not provided and &quot;</span>
                    <span class="s2">&quot;phenotypes has no index to retrieve pheno_samples from.&quot;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pheno_samples</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">phenotype_ID</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">((</span><span class="s2">&quot;phenotype_ID are not provided and phenotypes &quot;</span>
                    <span class="s2">&quot;has no column names to retrieve phenotype_ID from.&quot;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">phenotype_ID</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number of samples in phenotypes (</span><span class="si">{}</span><span class="s1">) does &#39;</span>
                <span class="s1">&#39;not match number of sample IDs (</span><span class="si">{}</span><span class="s1">) provided&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="k">if</span> <span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number phenotypes (</span><span class="si">{}</span><span class="s1">) does not match &#39;</span>
                <span class="s1">&#39;number of phenotype IDs (</span><span class="si">{}</span><span class="s1">) provided&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Duplicate sample names in phenotypes&quot;</span><span class="p">)</span>    
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Duplicate trait names in phenotypes&quot;</span><span class="p">)</span>    
        
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="p">,</span>
            <span class="n">columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">)</span></div>

<div class="viewcode-block" id="InputData.addCovariates"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.addCovariates">[docs]</a>    <span class="k">def</span> <span class="nf">addCovariates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">covariates</span><span class="p">,</span> <span class="n">covs_samples</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add [`N` x `K`] covariate data with [`N`] samples and [`K`] covariates</span>
<span class="sd">        to InputData instance.</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            covariates (array-like):</span>
<span class="sd">                [`N x `K`] covariate matrix of `N` individuals and `K`</span>
<span class="sd">                covariates; if pandas.DataFrame with covs_samples as index, </span>
<span class="sd">                covs_samples do not have to specified separately.</span>
<span class="sd">            covs_samples (array-like):</span>
<span class="sd">                [`N`] sample ID</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.covariates** (pd.DataFrame):</span>
<span class="sd">                  [`N` x `K`] covariates matrix</span>
<span class="sd">                - **self.covs_samples** (np.array):</span>
<span class="sd">                  [`N`] sample IDs</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; import numpy as np</span>
<span class="sd">                &gt;&gt;&gt; import pandas as pd</span>
<span class="sd">                &gt;&gt;&gt; covariates = [(1,2,4),(1,1,6),(0,4,8)]</span>
<span class="sd">                &gt;&gt;&gt; covs_samples = [&#39;S1&#39;,&#39;S2&#39;, &#39;S3&#39;]</span>
<span class="sd">                &gt;&gt;&gt; covariates = pd.DataFrame(covariates, index=covs_samples)</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addCovariates(covariates = covariates,</span>
<span class="sd">                ...     covs_samples = covs_samples)</span>
<span class="sd">                &gt;&gt;&gt; print indata.covariates.shape</span>
<span class="sd">                (3, 3)</span>
<span class="sd">                &gt;&gt;&gt; print indata.covs_samples.shape</span>
<span class="sd">                (3,)</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">covs_samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">covariates</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">((</span><span class="s2">&quot;covs_samples are not provided and &quot;</span>
                    <span class="s2">&quot;covariates has no index to retrieve covs_samples &quot;</span>
                    <span class="s2">&quot;from.&quot;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">covs_samples</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">covariates</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number of samples in covariates (</span><span class="si">{}</span><span class="s1">) does &#39;</span>
                <span class="s1">&#39;not match number of sample IDs (</span><span class="si">{}</span><span class="s1">) provided&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">covariates</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Duplicate sample names in covariates&quot;</span><span class="p">)</span>    
        <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">covariates</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span><span class="p">)</span></div>

<div class="viewcode-block" id="InputData.addRelatedness"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.addRelatedness">[docs]</a>    <span class="k">def</span> <span class="nf">addRelatedness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">relatedness</span><span class="p">,</span> <span class="n">relatedness_samples</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add [`N` x `N`] pairwise relatedness estimates of [`N`] samples to the</span>
<span class="sd">        InputData instance</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            relatedness (array-like):</span>
<span class="sd">                [`N x `N`] relatedness matrix of `N` individuals;</span>
<span class="sd">                if pandas.DataFrame with relatedness_samples as index,</span>
<span class="sd">                relatedness_samples do not have to specified separately.</span>
<span class="sd">            relatedness_samples (array-like):</span>
<span class="sd">                [`N`] sample IDs</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.relatedness** (pd.DataFrame):</span>
<span class="sd">                  [`N` x `N`] relatedness matrix</span>
<span class="sd">                - **self.relatedness_samples** (np.array):</span>
<span class="sd">                  [`N`] sample IDs</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; import numpy</span>
<span class="sd">                &gt;&gt;&gt; import pandas as pd</span>
<span class="sd">                &gt;&gt;&gt; from numpy.random import RandomState</span>
<span class="sd">                &gt;&gt;&gt; from numpy.linalg import cholesky as chol</span>
<span class="sd">                &gt;&gt;&gt; random = RandomState(5)</span>
<span class="sd">                &gt;&gt;&gt; N = 100</span>
<span class="sd">                &gt;&gt;&gt; SNP = 1000</span>
<span class="sd">                &gt;&gt;&gt; X = (random.rand(N, SNP) &lt; 0.3).astype(float)</span>
<span class="sd">                &gt;&gt;&gt; relatedness = numpy.dot(X, X.T)/float(SNP)</span>
<span class="sd">                &gt;&gt;&gt; relatedness_samples = numpy.array(</span>
<span class="sd">                ...     [&#39;S{}&#39;.format(x+1) for x in range(N)])</span>
<span class="sd">                &gt;&gt;&gt; relatedness = pd.DataFrame(relatedness,</span>
<span class="sd">                ...     index=relatedness_samples)</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addRelatedness(relatedness = relatedness)</span>
<span class="sd">                &gt;&gt;&gt; print indata.relatedness.shape</span>
<span class="sd">                (100, 100)</span>
<span class="sd">                &gt;&gt;&gt; print indata.relatedness_samples.shape</span>
<span class="sd">                (100,)</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">relatedness_samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">relatedness</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">((</span><span class="s2">&quot;relatedness_samples are not provided and &quot;</span>
                    <span class="s2">&quot;relatedness has no index to retrieve relatedness_samples &quot;</span>
                    <span class="s2">&quot;from&quot;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">relatedness_samples</span><span class="p">)</span>
        <span class="n">rel</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">rel</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">rel</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">FormatError</span><span class="p">((</span><span class="s1">&#39;Relatedness has to be a square matrix, but &#39;</span>
                <span class="s1">&#39;number of rows </span><span class="si">{}</span><span class="s1"> is not equal to number of columns &#39;</span>
                <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">rel</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">rel</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rel</span><span class="p">)</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rel</span><span class="p">)</span><span class="o">.</span><span class="n">T</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Relatedness matrix is not symmetric&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_is_positive_definite</span><span class="p">(</span><span class="n">rel</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Relatedness matrix is not positive-semi definite&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">rel</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number of samples in relatedness (</span><span class="si">{}</span><span class="s1">) does &#39;</span>
                    <span class="s1">&#39;not match number of sample IDs (</span><span class="si">{}</span><span class="s1">) provided&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">rel</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Duplicate sample names in relatedness&quot;</span><span class="p">)</span>    
        <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span><span class="p">)</span></div>

<div class="viewcode-block" id="InputData.addGenotypes"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.addGenotypes">[docs]</a>    <span class="k">def</span> <span class="nf">addGenotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">,</span> <span class="n">geno_samples</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
            <span class="n">genotypes_info</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add [`N` x `NrSNP`] genotype array of [`N`] samples and [`NrSNP`]</span>
<span class="sd">        genotypes, [`N`] array of sample IDs and [`NrSNP` x 2] dataframe of</span>
<span class="sd">        genotype description to InputData instance.</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            genotypes (array-like):</span>
<span class="sd">                [`N` x `NrSNP`] genotype array of [`N`] samples and [`NrSNP`]</span>
<span class="sd">                genotypes; if pandas.DataFrame with geno_samples as index,</span>
<span class="sd">                geno_samples do not have to specified separately.</span>
<span class="sd">            geno_samples (array-like):</span>
<span class="sd">                [`N`] vector of `N` sample IDs</span>
<span class="sd">            genotypes_info (dataframe):</span>
<span class="sd">                  [`NrSNPs` x 2] dataframe with columns &#39;chrom&#39; and &#39;pos&#39;, and</span>
<span class="sd">                  rsIDs as index</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.genotypes** (pd.DataFrame):</span>
<span class="sd">                  [`N` x `NrSNPs`] genotype matrix</span>
<span class="sd">                - **self.geno_samples** (np.array):</span>
<span class="sd">                  [`N`] sample IDs</span>
<span class="sd">                - **self.genotypes_info** (pd.DataFrame):</span>
<span class="sd">                  [`NrSNPs` x 2] dataframe with columns &#39;chrom&#39; and &#39;pos&#39;, and</span>
<span class="sd">                  rsIDs as index</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import reader</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; data = reader.ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_geno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                                &#39;io/test/data/genotypes.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getGenotypes(file_genotypes=file_geno)</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addGenotypes(genotypes=data.genotypes,</span>
<span class="sd">                ...                     genotypes_info=data.genotypes_info)</span>
<span class="sd">                &gt;&gt;&gt; indata.geno_samples[:5]</span>
<span class="sd">                array([&#39;ID_1&#39;, &#39;ID_2&#39;, &#39;ID_3&#39;, &#39;ID_4&#39;, &#39;ID_5&#39;], dtype=object)</span>
<span class="sd">                &gt;&gt;&gt; indata.genotypes.shape</span>
<span class="sd">                (1000, 20)</span>
<span class="sd">                &gt;&gt;&gt; indata.genotypes.values[:5,:5]</span>
<span class="sd">                array([[0., 0., 0., 0., 0.],</span>
<span class="sd">                       [0., 0., 0., 0., 0.],</span>
<span class="sd">                       [0., 0., 0., 0., 0.],</span>
<span class="sd">                       [2., 1., 0., 0., 0.],</span>
<span class="sd">                       [1., 0., 0., 0., 0.]])</span>
<span class="sd">                &gt;&gt;&gt; indata.genotypes_info[:5]</span>
<span class="sd">                           chrom       pos</span>
<span class="sd">                rs1601111      3  88905003</span>
<span class="sd">                rs13270638     8  20286021</span>
<span class="sd">                rs75132935     8  76564608</span>
<span class="sd">                rs72668606     8  79733124</span>
<span class="sd">                rs55770986     7   2087823</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">geno_samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotypes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">((</span><span class="s2">&quot;geno_samples are not provided and genotypes &quot;</span>
                    <span class="s2">&quot;has no index to retrieve geno_samples from&quot;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">geno_samples</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">genotypes_info</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">MissingInput</span><span class="p">((</span><span class="s1">&#39;Genotype info has to be specified via &#39;</span>
                    <span class="s1">&#39;genotypes_info&#39;</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span> <span class="o">=</span> <span class="n">genotypes_info</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number of samples in genotypes (</span><span class="si">{}</span><span class="s1">) does &#39;</span>
                <span class="s1">&#39;not match number of sample IDs (</span><span class="si">{}</span><span class="s1">) provided&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number of genotypes in genotypes (</span><span class="si">{}</span><span class="s1">) does &#39;</span>
                <span class="s1">&#39;not match number of genotypes in genotypes_info (</span><span class="si">{}</span><span class="s1">)&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Duplicate sample names in genotypes&quot;</span><span class="p">)</span>    </div>

<div class="viewcode-block" id="InputData.addVarianceComponents"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.addVarianceComponents">[docs]</a>    <span class="k">def</span> <span class="nf">addVarianceComponents</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">Cg</span><span class="p">,</span> <span class="n">Cn</span><span class="p">,):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add [`P` x `P`] matrices of [`P`] trait covariance estimates</span>
<span class="sd">        of the genetic trait variance component (Cg) and the non-genetic (noise)</span>
<span class="sd">        variance component (Cn) to InputData instance</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            Cg (array-like):</span>
<span class="sd">                [`P x `P`] matrix of `P` trait covariance estimates of the</span>
<span class="sd">                genetic trait covaraince component</span>
<span class="sd">            Cn (array-like):</span>
<span class="sd">                [`P x `P`] matrix of `P` trait covariance estimates of the</span>
<span class="sd">                non-genetic (noise) trait covaraince component</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.Cg** (np.array):</span>
<span class="sd">                  [`P x `P`] matrix of `P` trait covariance estimates of the</span>
<span class="sd">                  genetic trait covariance component</span>
<span class="sd">                - **self.Cn** (np.array):</span>
<span class="sd">                  [`P x `P`] matrix of `P` trait covariance estimates of the</span>
<span class="sd">                  non-genetic trait covaraince component</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import reader</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; import numpy as np</span>
<span class="sd">                &gt;&gt;&gt; from numpy.random import RandomState</span>
<span class="sd">                &gt;&gt;&gt; from numpy.linalg import cholesky as chol</span>
<span class="sd">                &gt;&gt;&gt; data = reader.ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_pheno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/pheno.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getPhenotypes(file_pheno=file_pheno)</span>
<span class="sd">                &gt;&gt;&gt; file_Cg = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/Cg.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; file_Cn = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/Cn.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getVarianceComponents(file_Cg=file_Cg,</span>
<span class="sd">                ...                            file_Cn=file_Cn)</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addPhenotypes(phenotypes = data.phenotypes)</span>
<span class="sd">                &gt;&gt;&gt; indata.addVarianceComponents(Cg = data.Cg, Cn=data.Cn)</span>
<span class="sd">                &gt;&gt;&gt; print indata.Cg.shape</span>
<span class="sd">                (10, 10)</span>
<span class="sd">                &gt;&gt;&gt; print indata.Cg.shape</span>
<span class="sd">                (10, 10)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">Cg</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">Cn</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">((</span><span class="s1">&#39;Phenotypes have to be added before Cg/Cn &#39;</span>
                        <span class="s1">&#39;can be added&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">Cg</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">Cn</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">((</span><span class="s1">&#39;Cg has to be a square matrix, but &#39;</span>
                    <span class="s1">&#39;number of rows </span><span class="si">{}</span><span class="s1"> is not equal to number of columns &#39;</span>
                    <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Cg</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="o">.</span><span class="n">T</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Cg is not symmetric&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_is_positive_definite</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Cg is not positive-semi definite&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number of traits in Cg (</span><span class="si">{}</span><span class="s1">) does &#39;</span>
                    <span class="s1">&#39;not match number of traits (</span><span class="si">{}</span><span class="s1">) in phenotypes&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">((</span><span class="s1">&#39;Cn has to be a square matrix, but &#39;</span>
                    <span class="s1">&#39;number of rows </span><span class="si">{}</span><span class="s1"> is not equal to number of columns &#39;</span>
                    <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Cn</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="o">.</span><span class="n">T</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Cn is not symmetric&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_is_positive_definite</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Cn is not positive-semi definite&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number of traits in Cn (</span><span class="si">{}</span><span class="s1">) does &#39;</span>
                    <span class="s1">&#39;not match number of traits (</span><span class="si">{}</span><span class="s1">) in phenotypes&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span></div>

<div class="viewcode-block" id="InputData.addPCs"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.addPCs">[docs]</a>    <span class="k">def</span> <span class="nf">addPCs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pcs</span><span class="p">,</span> <span class="n">pc_samples</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add [`N` x `PC`] matrix of [`PC`] principal components from the</span>
<span class="sd">        genotypes of [`N`] samples to InputData instance.</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            pcs (array-like):</span>
<span class="sd">                [`N x `PCs`] principal component matrix of `N` individuals and</span>
<span class="sd">                `PCs` principal components; if pandas.DataFrame with pc_samples</span>
<span class="sd">                as index, covs_samples do not have to specified separately.</span>
<span class="sd">            pc_samples (array-like):</span>
<span class="sd">                [`N`] sample IDs</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.pcs** (pd.DataFrame):</span>
<span class="sd">                  [`N` x `PCs`] principal component matrix</span>
<span class="sd">                - **self.pc_samples** (np.array):</span>
<span class="sd">                  [`N`] sample IDs</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import reader</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; data = reader.ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_pcs = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/pcs.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getPCs(file_pcs=file_pcs, nrpcs=10, delim=&quot; &quot;)</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addPCs(pcs = data.pcs)</span>
<span class="sd">                &gt;&gt;&gt; print indata.pcs.shape</span>
<span class="sd">                (1000, 10)</span>
<span class="sd">                &gt;&gt;&gt; print indata.pc_samples.shape</span>
<span class="sd">                (1000,)</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">pc_samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pcs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">((</span><span class="s2">&quot;pc_samples are not provided and pcs has &quot;</span>
                    <span class="s2">&quot;no index to retrieve pc_samples from&quot;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pc_samples</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Number of samples in pcs (</span><span class="si">{}</span><span class="s1">) does&#39;</span>
                    <span class="s1">&#39;not match number of sample IDs (</span><span class="si">{}</span><span class="s1">) provided&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> 
                        <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pc_samples</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Duplicate sample names for principle components&quot;</span><span class="p">)</span>    
        <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pcs</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span><span class="p">)</span></div>


<div class="viewcode-block" id="InputData.subsetTraits"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.subsetTraits">[docs]</a>    <span class="k">def</span> <span class="nf">subsetTraits</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">traitlist</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Limit analysis to specific subset of traits</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            traitlist (array-like):</span>
<span class="sd">                array of trait numbers to select from phenotypes</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.traitlist** (list):</span>
<span class="sd">                  of [`t`] trait numbers (int) to choose for analysis</span>
<span class="sd">                - **self.phenotypes** (pd.DataFrame):</span>
<span class="sd">                  reduced set of [`N` x `t`] phenotypes</span>
<span class="sd">                - **self.phenotype.ID** (np.array):</span>
<span class="sd">                  reduced set of [`t`] phenotype IDs</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>

<span class="sd">            .. doctest::</span>
<span class="sd">            </span>
<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.reader import ReadData</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.input import InputData</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.utils import file_type</span>
<span class="sd">                &gt;&gt;&gt; data = ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_pheno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                                &#39;io/test/data/pheno.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getPhenotypes(file_pheno=file_pheno)</span>
<span class="sd">                &gt;&gt;&gt; traitlist = data.getTraitSubset(traitstring=&quot;1-3,5&quot;)</span>
<span class="sd">                &gt;&gt;&gt; indata = InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addPhenotypes(phenotypes = data.phenotypes)</span>
<span class="sd">                &gt;&gt;&gt; print indata.phenotypes.shape</span>
<span class="sd">                (1000, 10)</span>
<span class="sd">                &gt;&gt;&gt; print indata.phenotype_ID.shape</span>
<span class="sd">                (10,)</span>
<span class="sd">                &gt;&gt;&gt; indata.subsetTraits(traitlist=traitlist)</span>
<span class="sd">                &gt;&gt;&gt; print indata.phenotypes.shape</span>
<span class="sd">                (1000, 4)</span>
<span class="sd">                &gt;&gt;&gt; print indata.phenotype_ID.shape</span>
<span class="sd">                (4,)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">traitlist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">traitlist</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">traitlist</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">traitlist</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Duplicate trait names in traitlist&quot;</span><span class="p">)</span>    
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="bp">self</span><span class="o">.</span><span class="n">traitlist</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">traitlist</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Selected trait number </span><span class="si">{}</span><span class="s1"> is greater &#39;</span>
                    <span class="s1">&#39;than number of phenotypes provided </span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">traitlist</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> 
                        <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span></div>

<div class="viewcode-block" id="InputData.commonSamples"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.commonSamples">[docs]</a>    <span class="k">def</span> <span class="nf">commonSamples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samplelist</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get [`M]` common samples out of phenotype, relatedness and optional</span>
<span class="sd">        covariates with [`N`] samples (if all samples present in all datasets</span>
<span class="sd">        [`M`] = [`N`]) and ensure that samples are in same order.</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            samplelist (array-like):</span>
<span class="sd">                array of sample IDs to select from data</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.phenotypes** (pd.DataFrame):</span>
<span class="sd">                  [`M` x `P`] phenotype matrix</span>
<span class="sd">                - **self.pheno_samples** (np.array):</span>
<span class="sd">                  [`M`] sample IDs</span>
<span class="sd">                - **self.relatedness** (pd.DataFrame):</span>
<span class="sd">                  [`M x M`] relatedness matrix</span>
<span class="sd">                - **self.relatedness_samples** (np.array):</span>
<span class="sd">                  [`M`] sample IDs of relatedness matrix</span>
<span class="sd">                - **self.covariates** (pd.DataFrame):</span>
<span class="sd">                  [`M` x `K`] covariates matrix</span>
<span class="sd">                - **self.covs_samples** (np.array):</span>
<span class="sd">                  [`M`] sample IDs</span>
<span class="sd">                - **self.genotypes** (pd.DataFrame):</span>
<span class="sd">                  [`M` x `NrSNPs`] genotypes matrix</span>
<span class="sd">                - **self.geno_samples** (np.array):</span>
<span class="sd">                  [`M`] sample IDs</span>
<span class="sd">                - **self.pcs** (pd.DataFrame):</span>
<span class="sd">                  [`M` x `PCs`] principal component matrix</span>
<span class="sd">                - **self.pc_samples** (np.array):</span>
<span class="sd">                  [`M`] sample IDs</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; import numpy as np</span>
<span class="sd">                &gt;&gt;&gt; import pandas as pd</span>
<span class="sd">                &gt;&gt;&gt; from numpy.random import RandomState</span>
<span class="sd">                &gt;&gt;&gt; from numpy.linalg import cholesky as chol</span>
<span class="sd">                &gt;&gt;&gt; random = RandomState(5)</span>
<span class="sd">                &gt;&gt;&gt; P = 2</span>
<span class="sd">                &gt;&gt;&gt; K = 4</span>
<span class="sd">                &gt;&gt;&gt; N = 10</span>
<span class="sd">                &gt;&gt;&gt; SNP = 1000</span>
<span class="sd">                &gt;&gt;&gt; pheno = random.normal(0,1, (N, P))</span>
<span class="sd">                &gt;&gt;&gt; pheno_samples = np.array([&#39;S{}&#39;.format(x+4)</span>
<span class="sd">                ...     for x in range(N)])</span>
<span class="sd">                &gt;&gt;&gt; phenotype_ID = np.array([&#39;ID{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(P)])</span>
<span class="sd">                &gt;&gt;&gt; phenotypes = pd.DataFrame(pheno, index=pheno_samples,</span>
<span class="sd">                ...     columns=phenotype_ID)</span>
<span class="sd">                &gt;&gt;&gt; X = (random.rand(N, SNP) &lt; 0.3).astype(float)</span>
<span class="sd">                &gt;&gt;&gt; relatedness = np.dot(X, X.T)/float(SNP)</span>
<span class="sd">                &gt;&gt;&gt; relatedness_samples = np.array([&#39;S{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(N)])</span>
<span class="sd">                &gt;&gt;&gt; covariates = random.normal(0,1, (N-2, K))</span>
<span class="sd">                &gt;&gt;&gt; covs_samples = np.array([&#39;S{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(N-2)])</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addPhenotypes(phenotypes = pheno,</span>
<span class="sd">                ...                      pheno_samples = pheno_samples,</span>
<span class="sd">                ...                      phenotype_ID = phenotype_ID)</span>
<span class="sd">                &gt;&gt;&gt; indata.addRelatedness(relatedness = relatedness,</span>
<span class="sd">                ...                  relatedness_samples = relatedness_samples)</span>
<span class="sd">                &gt;&gt;&gt; indata.addCovariates(covariates = covariates,</span>
<span class="sd">                ...                      covs_samples = covs_samples)</span>
<span class="sd">                &gt;&gt;&gt; indata.covariates.shape</span>
<span class="sd">                (8, 4)</span>
<span class="sd">                &gt;&gt;&gt; indata.phenotypes.shape</span>
<span class="sd">                (10, 2)</span>
<span class="sd">                &gt;&gt;&gt; indata.relatedness.shape</span>
<span class="sd">                (10, 10)</span>
<span class="sd">                &gt;&gt;&gt; indata.commonSamples(samplelist=[&quot;S4&quot;, &quot;S6&quot;, &quot;S5&quot;])</span>
<span class="sd">                &gt;&gt;&gt; indata.covariates.shape</span>
<span class="sd">                (3, 4)</span>
<span class="sd">                &gt;&gt;&gt; indata.phenotypes.shape</span>
<span class="sd">                (3, 2)</span>
<span class="sd">                &gt;&gt;&gt; indata.relatedness.shape</span>
<span class="sd">                (3, 3)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">test_pheno_relatedness</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="p">,</span> 
                    <span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">test_pheno_relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;No common samples between phenotypes and &#39;</span>
                        <span class="s1">&#39;relatedness estimates&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">test_pheno_relatedness</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">test_pheno_geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="p">,</span> 
                    <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">test_pheno_geno</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;No common samples between phenotypes,&#39;</span>
                        <span class="s1">&#39;and genotypes&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">test_pheno_geno</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">test_pheno_covs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="p">,</span> 
                    <span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">test_pheno_covs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;No common samples between phenotypes, &#39;</span>
                        <span class="s1">&#39;and covariates&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">test_pheno_covs</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">test_pheno_pcs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span><span class="p">,</span> 
                    <span class="bp">self</span><span class="o">.</span><span class="n">pcs_samples</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">test_pheno_pcs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;No common samples between phenotypes,&#39;</span>
                        <span class="s1">&#39;and pcs&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">test_pheno_pcs</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">samplelist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">samplelist</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">samplelist</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Duplicate sample names in samplelist&quot;</span><span class="p">)</span>
            <span class="n">test_samples_samplelist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">samplelist</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">test_samples_samplelist</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;No samples between common samples in, &#39;</span>
                        <span class="s1">&#39;datasets and samplelist&#39;</span><span class="p">))</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">test_samples_samplelist</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">samplelist</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Not all Ids in samplelist are contained &#39;</span>
                    <span class="s1">&#39;in common samples from provided datasets&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">samplelist</span>
    
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,:]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,:]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,:]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,:]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">,:]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pcs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span></div>

<div class="viewcode-block" id="InputData.regress"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.regress">[docs]</a>    <span class="k">def</span> <span class="nf">regress</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Regress out covariates (optional).</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.phenotypes** (np.array):</span>
<span class="sd">                  [`M` x `P`] phenotype matrix of residuals of linear model</span>
<span class="sd">                - **self.covariates**:</span>
<span class="sd">                  None</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; import numpy as np</span>
<span class="sd">                &gt;&gt;&gt; from numpy.random import RandomState</span>
<span class="sd">                &gt;&gt;&gt; from numpy.linalg import cholesky as chol</span>
<span class="sd">                &gt;&gt;&gt; random = RandomState(5)</span>
<span class="sd">                &gt;&gt;&gt; P = 5</span>
<span class="sd">                &gt;&gt;&gt; K = 4</span>
<span class="sd">                &gt;&gt;&gt; N = 100</span>
<span class="sd">                &gt;&gt;&gt; pheno = random.normal(0,1, (N, P))</span>
<span class="sd">                &gt;&gt;&gt; pheno_samples = np.array([&#39;S{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(N)])</span>
<span class="sd">                &gt;&gt;&gt; phenotype_ID = np.array([&#39;ID{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(P)])</span>
<span class="sd">                &gt;&gt;&gt; covariates = random.normal(0,1, (N, K))</span>
<span class="sd">                &gt;&gt;&gt; covs_samples = np.array([&#39;S{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(N)])</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addPhenotypes(phenotypes = pheno,</span>
<span class="sd">                ...                      pheno_samples = pheno_samples,</span>
<span class="sd">                ...                      phenotype_ID = phenotype_ID)</span>
<span class="sd">                &gt;&gt;&gt; indata.addCovariates(covariates = covariates,</span>
<span class="sd">                ...                      covs_samples = covs_samples)</span>
<span class="sd">                &gt;&gt;&gt; indata.phenotypes.values[:3, :3]</span>
<span class="sd">                array([[ 0.44122749, -0.33087015,  2.43077119],</span>
<span class="sd">                       [ 1.58248112, -0.9092324 , -0.59163666],</span>
<span class="sd">                       [-1.19276461, -0.20487651, -0.35882895]])</span>
<span class="sd">                &gt;&gt;&gt; indata.regress()</span>
<span class="sd">                &gt;&gt;&gt; indata.phenotypes.values[:3, :3]</span>
<span class="sd">                array([[ 0.34421705, -0.01470998,  2.25710966],</span>
<span class="sd">                       [ 1.69886647, -1.41756814, -0.55614649],</span>
<span class="sd">                       [-1.10700674, -0.66017713, -0.22201814]])</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">array_equal</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">DataMismatch</span><span class="p">((</span><span class="s1">&#39;Phenotype and covariate arrays are &#39;</span>
                    <span class="s1">&#39;identical&#39;</span><span class="p">))</span>
        <span class="n">verboseprint</span><span class="p">(</span><span class="s1">&#39;Regress covariates&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
        <span class="n">phenotypes</span> <span class="o">=</span> <span class="n">regressOut</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">),</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="InputData.transform"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.transform">[docs]</a>    <span class="k">def</span> <span class="nf">transform</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transform</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Transform phenotypes</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            transform (string):</span>
<span class="sd">                transformation method for phenotype data:</span>
<span class="sd">        </span>
<span class="sd">                    - scale:</span>
<span class="sd">                      mean center, divide by sd</span>
<span class="sd">                    - gaussian:</span>
<span class="sd">                      inverse normalisation</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.phenotypes** (np.array):</span>
<span class="sd">                  [`N` x `P`] (transformed) phenotype matrix</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; import numpy as np</span>
<span class="sd">                &gt;&gt;&gt; from numpy.random import RandomState</span>
<span class="sd">                &gt;&gt;&gt; from numpy.linalg import cholesky as chol</span>
<span class="sd">                &gt;&gt;&gt; random = RandomState(5)</span>
<span class="sd">                &gt;&gt;&gt; P = 5</span>
<span class="sd">                &gt;&gt;&gt; K = 4</span>
<span class="sd">                &gt;&gt;&gt; N = 100</span>
<span class="sd">                &gt;&gt;&gt; pheno = random.normal(0,1, (N, P))</span>
<span class="sd">                &gt;&gt;&gt; pheno_samples = np.array([&#39;S{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(N)])</span>
<span class="sd">                &gt;&gt;&gt; phenotype_ID = np.array([&#39;ID{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(P)])</span>
<span class="sd">                &gt;&gt;&gt; SNP = 1000</span>
<span class="sd">                &gt;&gt;&gt; X = (random.rand(N, SNP) &lt; 0.3).astype(float)</span>
<span class="sd">                &gt;&gt;&gt; relatedness = np.dot(X, X.T)/float(SNP)</span>
<span class="sd">                &gt;&gt;&gt; relatedness_samples = np.array([&#39;S{}&#39;.format(x+1)</span>
<span class="sd">                ...     for x in range(N)])</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addPhenotypes(phenotypes = pheno,</span>
<span class="sd">                ...                      pheno_samples = pheno_samples,</span>
<span class="sd">                ...                      phenotype_ID = phenotype_ID)</span>
<span class="sd">                &gt;&gt;&gt; indata.addRelatedness(relatedness = relatedness,</span>
<span class="sd">                ...                  relatedness_samples = relatedness_samples)</span>
<span class="sd">                &gt;&gt;&gt; indata.phenotypes.values[:3, :3]</span>
<span class="sd">                array([[ 0.44122749, -0.33087015,  2.43077119],</span>
<span class="sd">                       [ 1.58248112, -0.9092324 , -0.59163666],</span>
<span class="sd">                       [-1.19276461, -0.20487651, -0.35882895]])</span>
<span class="sd">                &gt;&gt;&gt; indata.transform(transform=&#39;gaussian&#39;)</span>
<span class="sd">                &gt;&gt;&gt; indata.phenotypes.values[:3, :3]</span>
<span class="sd">                array([[ 0.23799988, -0.11191464,  2.05785598],</span>
<span class="sd">                       [ 1.41041953, -0.81365681, -0.92217818],</span>
<span class="sd">                       [-1.55977999,  0.01240937, -0.62091817]])</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">transform</span> <span class="o">==</span> <span class="s1">&#39;scale&#39;</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s1">&#39;Use </span><span class="si">%s</span><span class="s1"> as transformation&#39;</span> <span class="o">%</span> <span class="n">transform</span><span class="p">,</span> 
                <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="n">phenotypes</span> <span class="o">=</span> <span class="n">scale</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">,</span>
                <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">transform</span> <span class="o">==</span> <span class="s1">&#39;gaussian&#39;</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s1">&#39;Use </span><span class="si">%s</span><span class="s1"> as transformation&#39;</span> <span class="o">%</span> <span class="n">transform</span><span class="p">,</span> 
                <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="n">phenotypes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">apply_along_axis</span><span class="p">(</span><span class="n">quantile_gaussianize</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> 
                    <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">,</span>
                <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">((</span><span class="s1">&#39;Possible transformation methods are: scale, &#39;</span>
                    <span class="s1">&#39;and gaussian but </span><span class="si">{}</span><span class="s1"> provided&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">transform</span><span class="p">))</span></div>

<div class="viewcode-block" id="InputData.standardiseGenotypes"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.standardiseGenotypes">[docs]</a>    <span class="k">def</span> <span class="nf">standardiseGenotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Standardise genotypes:</span>

<span class="sd">        .. math::</span>
<span class="sd">           w_{ij} = \frac{x_{ij} -2p_i}{\sqrt{2p_i (1-p_i)}} </span>
<span class="sd">        </span>
<span class="sd">        where :math:`x_{ij}` is the number of copies of the reference allele for </span>
<span class="sd">        the :math:`i` th SNP of the :math:`j` th individual and :math:`p_i` is </span>
<span class="sd">        the frequency of the reference allele (as described in `(Yang et al</span>
<span class="sd">        2011)</span>
<span class="sd">        &lt;http://www.pubmedcentral.nih.gov/articlerender.fcgi?artid=3014363&amp;tool=pmcentrez&amp;rendertype=abstract&gt;`_).</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.genotypes_sd** (numpy array):</span>
<span class="sd">                  [`N` x `NrSNP`] matrix of `NrSNP` standardised genotypes for</span>
<span class="sd">                  `N` samples.</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import reader</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.utils.utils import makeHardCalledGenotypes</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.utils.utils import AlleleFrequencies</span>
<span class="sd">                &gt;&gt;&gt; data = reader.ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_geno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                                &#39;io/test/data/genotypes.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getGenotypes(file_genotypes=file_geno)</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addGenotypes(genotypes=data.genotypes,</span>
<span class="sd">                ...                     genotypes_info=data.genotypes_info)</span>
<span class="sd">                &gt;&gt;&gt; geno_sd = indata.standardiseGenotypes()</span>
<span class="sd">                &gt;&gt;&gt; geno_sd.iloc[:5,:3]</span>
<span class="sd">                             0         1       2</span>
<span class="sd">                ID_1 -2.201123 -2.141970 -8.9622</span>
<span class="sd">                ID_2 -2.201123 -2.141970 -8.9622</span>
<span class="sd">                ID_3 -2.201123 -2.141970 -8.9622</span>
<span class="sd">                ID_4  0.908627 -0.604125 -8.9622</span>
<span class="sd">                ID_5 -0.646248 -2.141970 -8.9622</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_sd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">snp</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
            <span class="n">p</span><span class="p">,</span> <span class="n">q</span> <span class="o">=</span> <span class="n">AlleleFrequencies</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">snp</span><span class="p">])</span>
            <span class="n">var_snp</span> <span class="o">=</span> <span class="n">sqrt</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">p</span> <span class="o">*</span> <span class="n">q</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">snp</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_sd</span><span class="p">[</span><span class="n">n</span><span class="p">,</span> <span class="n">snp</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="p">)[</span><span class="n">n</span><span class="p">,</span> <span class="n">snp</span><span class="p">]</span> <span class="o">-</span>
                    <span class="mi">2</span> <span class="o">*</span> <span class="n">q</span><span class="p">)</span> <span class="o">/</span> <span class="n">var_snp</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_sd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_sd</span><span class="p">,</span>
                <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_sd</span></div>

<div class="viewcode-block" id="InputData.getAlleleFrequencies"><a class="viewcode-back" href="../../../input.html#limmbo.io.input.InputData.getAlleleFrequencies">[docs]</a>    <span class="k">def</span> <span class="nf">getAlleleFrequencies</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Compute allele frequencies of genotypes.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the InputData instance:</span>
<span class="sd">        </span>
<span class="sd">                - **self.freqs** (pandas DataFrame):</span>
<span class="sd">                  [`NrSNP` x `2`] matrix of alt and ref allele frequencies; </span>
<span class="sd">                  index: snp IDs</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import reader</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import input</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.utils.utils import makeHardCalledGenotypes</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.utils.utils import AlleleFrequencies</span>
<span class="sd">                &gt;&gt;&gt; data = reader.ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_geno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                                &#39;io/test/data/genotypes.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getGenotypes(file_genotypes=file_geno)</span>
<span class="sd">                &gt;&gt;&gt; indata = input.InputData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; indata.addGenotypes(genotypes=data.genotypes,</span>
<span class="sd">                ...                     genotypes_info=data.genotypes_info,</span>
<span class="sd">                ...                     geno_samples=data.geno_samples)</span>
<span class="sd">                &gt;&gt;&gt; freqs = indata.getAlleleFrequencies()</span>
<span class="sd">                &gt;&gt;&gt; freqs.iloc[:5,:]</span>
<span class="sd">                                   p         q</span>
<span class="sd">                rs1601111   0.292186  0.707814</span>
<span class="sd">                rs13270638  0.303581  0.696419</span>
<span class="sd">                rs75132935  0.024295  0.975705</span>
<span class="sd">                rs72668606  0.119091  0.880909</span>
<span class="sd">                rs55770986  0.169338  0.830662</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">verboseprint</span><span class="p">(</span><span class="s1">&#39;Get allele frequencies of </span><span class="si">%s</span><span class="s1"> snps&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">freqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">2</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">snp</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">freqs</span><span class="p">[</span><span class="n">snp</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">freqs</span><span class="p">[</span><span class="n">snp</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">AlleleFrequencies</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="p">)[:,</span> <span class="n">snp</span><span class="p">])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">freqs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">freqs</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> 
                <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="s1">&#39;q&#39;</span><span class="p">])</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">freqs</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_is_positive_definite</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span> 
            <span class="n">chol_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">cholesky</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
            <span class="k">return</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">LinAlgError</span><span class="p">:</span>
            <span class="k">return</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Hannah V. Meyer.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'b',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>