

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>limmbo.io.parser &mdash; LiMMBo b documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="LiMMBo b documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> LiMMBo
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../input.html">Data input</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../vd.html">Variance decomposition</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gwas.html">Association analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cmd.html">Command-line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">Additional functions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">LiMMBo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>limmbo.io.parser</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for limmbo.io.parser</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">argparse</span>

<div class="viewcode-block" id="getGWASargs"><a class="viewcode-back" href="../../../input.html#limmbo.io.parser.getGWASargs">[docs]</a><span class="k">def</span> <span class="nf">getGWASargs</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">prog</span><span class="o">=</span><span class="s2">&quot;runAssociation&quot;</span><span class="p">,</span> 
            <span class="n">description</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Models the association between phenotypes and &#39;</span>
                <span class="s1">&#39;genotypes, accepting additional covariates and parameters &#39;</span>
                <span class="s1">&#39;to account for population structure and relatedness between &#39;</span>
                <span class="s1">&#39;samples. Users can choose between single-trait and multi-&#39;</span>
                <span class="s1">&#39;trait models, simple linear or linear mixed model set-ups.&#39;</span><span class="p">))</span>
    <span class="n">required</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Basic required arguments&#39;</span><span class="p">)</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-p&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--file_pheno&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_pheno&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] to [(N+1) x (P+1)] .csv file of [P] &#39;</span> 
            <span class="s1">&#39;phenotypes with [N] samples (first column: sample IDs, first &#39;</span>
            <span class="s1">&#39;row: phenotype IDs). Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--pheno_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;pheno_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of phenotype file.  g &#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-g&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;--file_geno&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_genotypes&quot;</span><span class="p">,</span> 
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Genotype file: either [S x N].csv file (first column: SNP id, &#39;</span>
            <span class="s1">&#39;first row: sample IDs) or  `plink formated &lt;https://www.cog-genomics.org/plink/1.9/input&gt;`_ genotypes &#39;</span>
             <span class="s1">&#39;(.bim/.fam/.bed). Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--geno_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;genotypes_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of genotype file (if not in `plink format &lt;https://www.cog-genomics.org/plink/1.9/input&gt;`_). &#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-o&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--outdir&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;outdir&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] of output directory; user needs writing &#39;</span>
            <span class="s1">&#39;permission. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">mode</span> <span class="o">=</span> <span class="n">required</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">(</span><span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mode</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-st&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--singletrait&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;singletrait&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Set flag to conduct a single-trait association analyses&#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    
    <span class="n">mode</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-mt&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--multitrait&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;multitrait&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Set flag to conduct a multi-trait association analyses&#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">setup</span> <span class="o">=</span> <span class="n">required</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">(</span><span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">setup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-lm&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--lm&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;lm&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Set flag to use a simple linear model for the association&#39;</span>
              <span class="s1">&#39;analysis&#39;</span><span class="p">))</span>
    <span class="n">setup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-lmm&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--lmm&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;lmm&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Set flag to use a linear mixed model for the association&#39;</span>
              <span class="s1">&#39;analysis&#39;</span><span class="p">))</span>
    
    <span class="n">output</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Output arguments&#39;</span><span class="p">)</span>
    <span class="n">output</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-n&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;--name&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Name (used for output file naming). Default: &#39;</span>
             <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">output</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-v&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--verbose&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;verbose&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;[bool]: should analysis progress be displayed. &#39;</span>
            <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    
    <span class="n">optionalfiles</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Optional files&#39;</span><span class="p">)</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-k&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--file_kinship&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_relatedness&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] to [N x (N+1)] file of kinship/relatedness &#39;</span>
            <span class="s1">&#39;matrix with [N] samples (first row: sample IDs). Required when &#39;</span>
            <span class="s1">&#39;--lmm/-lm. Default: &#39;</span>
            <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--kinship_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;relatedness_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of kinship file. &#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-cg&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;--file_cg&#39;</span><span class="p">,</span> 
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_cg&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Required for large phenotype sizes when --lmm/-lm; computed via &#39;</span>
             <span class="s1">&#39;runLiMMBo; specifies file name for genetic trait covariance &#39;</span>
             <span class="s1">&#39; matrix (rows: traits, columns: traits). Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--cg_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;cg_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of Cg file. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-cn&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;--file_cn&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_cn&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Required for large phenotype sizeswhen --lmm/-lm; computed via &#39;</span>
             <span class="s1">&#39;runLiMMBo; specifies file name for non-genetic trait covariance &#39;</span>
             <span class="s1">&#39; matrix (rows: traits, columns: traits). Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--cn_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;cn_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of Cn file. g Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-pcs&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;--file_pcs&#39;</span><span class="p">,</span> 
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_pcs&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path to [N x PCs] file of principal components from &#39;</span>
             <span class="s1">&#39;genotypes to be included as covariates (first column: &#39;</span>
             <span class="s1">&#39;sample IDs, first row: PC IDs); Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--pcs_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;pcs_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of PCs file. g Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-c&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--file_cov&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_covariates&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] to [(N+1) x C] file of covariates matrix with &#39;</span>
             <span class="s1">&#39;[N] samples and [K] covariates (first column: sample IDs, &#39;</span>
             <span class="s1">&#39;first row: phenotype IDs). Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--covariate_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;covariate_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of covariates file. g &#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>


    <span class="n">optional</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Optional association parameters&#39;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-adjustP&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--adjustP&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;adjustP&quot;</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;bonferroni&#39;</span><span class="p">,</span> <span class="s1">&#39;effective&#39;</span><span class="p">,</span> <span class="s1">&#39;None&#39;</span><span class="p">],</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Method to adjust single-trait p-values for&#39;</span>
              <span class="s1">&#39;multiple hypothesis testing when running&#39;</span>
              <span class="s1">&#39;multiple single-trait GWAS: bonferroni/effective number of &#39;</span>
              <span class="s1">&#39;tests `(Galwey,2009) &lt;http://onlinelibrary.wiley.com/doi/10.1002/gepi.20408/abstract&gt;`_ &#39;</span>
              <span class="s1">&#39;. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-nrpermutations&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--nrpermutations&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;nrpermutations&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Number of permutations for computing empirical p-values; &#39;</span>
            <span class="s1">&#39;1/nrpermutations is maximum level of testing for &#39;</span>
            <span class="s1">&#39;significance. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-fdr&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--fdr&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;fdr&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;FDR threshold for computing empirical FDR. Default: &#39;</span>
            <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-seed&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--seed&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;seed&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">256</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Seed [int] to inittiate random number generation for &#39;</span>
            <span class="s1">&#39;permutations. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">),</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

    <span class="n">data</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Optional data processing parameters&#39;</span><span class="p">)</span>
    <span class="n">data</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-tr&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--transform_method&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;transform&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s1">&#39;scale&#39;</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;scale&#39;</span><span class="p">,</span> <span class="s1">&#39;gaussian&#39;</span><span class="p">],</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Choose type [string] of data preprocessing: scale (mean &#39;</span>
            <span class="s1">&#39;center, divide by sd) or gaussian (inverse normalise). &#39;</span> 
            <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">data</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-reg&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--reg_covariates&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;regress&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;[bool]: should covariates be regressed out? Default: &#39;</span>
            <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">subset</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Optional subsetting options&#39;</span><span class="p">)</span>
    <span class="n">subset</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-traitset&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--traitset&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;traitstring&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Comma- (for list of traits) or hyphen- (for trait range) or &#39;</span>
        <span class="s1">&#39;comma and hyphen-separated list [string] of traits (trait &#39;</span>
        <span class="s1">&#39;columns) to choose; default: None (=all traits). Default: &#39;</span>
        <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    
    <span class="n">subset</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-nrpcs&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--nrpcs&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;nrpcs&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;First PCs to chose. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span> <span class="nb">int</span><span class="p">)</span>

    <span class="n">samples</span> <span class="o">=</span> <span class="n">subset</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">(</span><span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">samples</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--file_samplelist&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_samplelist&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] to file with samplelist for sample &#39;</span>
            <span class="s1">&#39;selection, with one sample ID per line. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    
    <span class="n">samples</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--samplelist&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;samplelist&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Comma-separated list [string] of samples IDs to restrict &#39;</span>
            <span class="s1">&#39;analysis to, e.g. ID1,ID2,ID5,ID9,ID10. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">plotting</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Plot arguments&#39;</span><span class="p">,</span> 
            <span class="s1">&#39;Arguments for depicting GWAS results as manhattan plot&#39;</span><span class="p">)</span>
    <span class="n">plotting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--plot&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;plot&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Set flag if results of association analysis should be depicted &#39;</span>
             <span class="s1">&#39;as manhattan and quantile-quantile plot&#39;</span><span class="p">))</span>
    <span class="n">plotting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-colourS&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--colourS&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;colourS&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s1">&#39;DarkBlue&#39;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Colour of significant points in manhattan plot&#39;</span><span class="p">))</span>
    <span class="n">plotting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-colourNS&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--colourNS&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;colourNS&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s1">&#39;Orange&#39;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Colour of non-significant points in manhattan plot&#39;</span><span class="p">))</span>
    <span class="n">plotting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-alphaS&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--alphaS&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;alphaS&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Transparency of significant points in manhattan plot&#39;</span><span class="p">))</span>
    <span class="n">plotting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-alphaNS&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--alphaNS&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;alphaNS&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Transparency of non-significant points in manhattan plot&#39;</span><span class="p">))</span>
    <span class="n">plotting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-thr&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--threshold&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;thr&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Significance threshold; when --fdr specified, empirical fdr used &#39;</span>
             <span class="s1">&#39;as threshold&#39;</span><span class="p">))</span>

    <span class="n">version</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Version&#39;</span><span class="p">)</span>
    <span class="n">version</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--version&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;version&#39;</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(prog)s</span><span class="s1"> 0.1.0&#39;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">parser</span></div>

<div class="viewcode-block" id="getVarianceEstimationArgs"><a class="viewcode-back" href="../../../input.html#limmbo.io.parser.getVarianceEstimationArgs">[docs]</a><span class="k">def</span> <span class="nf">getVarianceEstimationArgs</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">prog</span><span class="o">=</span><span class="s2">&quot;runVarianceEstimation&quot;</span><span class="p">,</span>
            <span class="n">description</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Estimates the genetic and non-genetic trait&#39;</span> 
                <span class="s1">&#39;covariance matrix parameters of a linear mixed model &#39;</span>
                <span class="s1">&#39;with random genetic and non-genetic effect via a &#39;</span>
                <span class="s1">&#39;bootstrapping-based approach.&#39;</span><span class="p">))</span>
    <span class="n">required</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Basic required arguments&#39;</span><span class="p">)</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-p&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--file_pheno&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_pheno&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] to [(N+1) x (P+1)] .csv file of [P] &#39;</span>
            <span class="s1">&#39;phenotypes with [N] samples (first column: sample IDs, first &#39;</span>
            <span class="s1">&#39;row: phenotype IDs). Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--pheno_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;pheno_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of phenotype file. &#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-k&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--file_kinship&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_relatedness&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] to [N x (N+1)] file of kinship/relatedness &#39;</span>
            <span class="s1">&#39;matrix with [N] samples (first row: sample IDs). Required when &#39;</span>
            <span class="s1">&#39;--lmm/-lm. Default: &#39;</span>
            <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--kinship_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;relatedness_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of kinship file. &#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">required</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-o&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--outdir&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;outdir&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] of output directory; user needs writing &#39;</span>
            <span class="s1">&#39;permission. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">optionalfiles</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Optional files&#39;</span><span class="p">)</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-c&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--file_cov&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_covariates&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] to [(N+1) x C] file of covariates matrix with &#39;</span>
             <span class="s1">&#39;[N] samples and [K] covariates (first column: sample IDs, &#39;</span>
             <span class="s1">&#39;first row: phenotype IDs). Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">optionalfiles</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--covariate_delim&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;covariate_delim&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Delimiter of covariates file. &#39;</span>
             <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">limmbo</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Bootstrapping parameters&#39;</span><span class="p">)</span>
    <span class="n">limmbo</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-seed&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--seed&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;seed&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">234</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;seed [int] used to generate bootstrap matrix. Default: &#39;</span>
            <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">),</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

    <span class="n">limmbo</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-sp&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--smallp&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;S&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Size [int] of phenotype subsamples used for variance &#39;</span>
            <span class="s1">&#39;decomposition. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">),</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">limmbo</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-r&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--runs&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;runs&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Total number [int] of bootstrap runs. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">limmbo</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--timing&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;timing&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;[bool]: should variance decomposition be timed. Default: &#39;</span>
            <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">limmbo</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--minCooccurrence&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;minCooccurrence&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Minimum count [int] of the pairwise sampling of any given &#39;</span>
            <span class="s1">&#39;trait pair. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">limmbo</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-i&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--iterations&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;iterations&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Number [int] of iterations for variance decomposition &#39;</span>
            <span class="s1">&#39;attempts. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">limmbo</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-cpus&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--cpus&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;cpus&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Number [int] of available CPUs for parallelisation of &#39;</span>
            <span class="s1">&#39;variance decomposition steps. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    
    <span class="n">data</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Optional data processing parameters&#39;</span><span class="p">)</span>
    <span class="n">data</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-tr&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--transform_method&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;transform&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s1">&#39;scale&#39;</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;scale&#39;</span><span class="p">,</span> <span class="s1">&#39;gaussian&#39;</span><span class="p">],</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Choose type [string] of data preprocessing: scale (mean &#39;</span>
            <span class="s1">&#39;center, divide by sd) or gaussian (inverse normalise). &#39;</span>
            <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="n">data</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-reg&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--reg_covariates&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;regress&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;[bool]: should covariates be regressed out? Default: &#39;</span>
            <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>


    <span class="n">subset</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Optional subsetting options&#39;</span><span class="p">)</span>
    <span class="n">subset</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-traitset&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--traitset&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;traitstring&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Comma- (for list of traits) or hyphen- (for trait range) or &#39;</span>
        <span class="s1">&#39;comma and hyphen-separated list [string] of traits (trait &#39;</span>
        <span class="s1">&#39;columns) to choose; default: None (=all traits). Default: &#39;</span>
        <span class="s1">&#39;</span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">samples</span> <span class="o">=</span> <span class="n">subset</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">(</span><span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">samples</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--file_samplelist&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;file_samplelist&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Path [string] to file with samplelist for sample &#39;</span>
            <span class="s1">&#39;selection, with one sample ID per line.Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">samples</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;--samplelist&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;samplelist&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Comma-separated list [string] of samples IDs to restrict &#39;</span>
            <span class="s1">&#39;analysis to, e.g. ID1,ID2,ID5,ID9,ID10. Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>

    <span class="n">output</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Output arguments&#39;</span><span class="p">)</span>
    <span class="n">output</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-dontSaveIntermediate&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--dontSaveIntermediate&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_false&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;intermediate&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;Set to suppress saving intermediate variance components. &#39;</span>
            <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    
    <span class="n">output</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s1">&#39;-v&#39;</span><span class="p">,</span>
        <span class="s1">&#39;--verbose&#39;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;verbose&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;[bool]: should analysis step description be printed. &#39;</span>
            <span class="s1">&#39;Default: </span><span class="si">%(default)s</span><span class="s1">&#39;</span><span class="p">))</span>
    
    <span class="n">version</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Version&#39;</span><span class="p">)</span>
    <span class="n">version</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--version&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;version&#39;</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(prog)s</span><span class="s1"> 0.1.0&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Hannah V. Meyer.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'b',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>