

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>limmbo.io.reader &mdash; LiMMBo b documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="LiMMBo b documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> LiMMBo
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../input.html">Data input</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../vd.html">Variance decomposition</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../gwas.html">Association analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cmd.html">Command-line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">Additional functions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">LiMMBo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>limmbo.io.reader</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for limmbo.io.reader</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">scipy</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">re</span>

<span class="kn">from</span> <span class="nn">limix.io</span> <span class="k">import</span> <span class="n">read_plink</span>

<span class="kn">from</span> <span class="nn">limmbo.utils.utils</span> <span class="k">import</span> <span class="n">verboseprint</span>
<span class="kn">from</span> <span class="nn">limmbo.io.utils</span> <span class="k">import</span> <span class="n">file_type</span>

<span class="k">class</span> <span class="nc">MissingInput</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Raised when no appropriate input is given&quot;&quot;&quot;</span>
    <span class="k">pass</span>


<span class="k">class</span> <span class="nc">FormatError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Raised when no appropriate input is given&quot;&quot;&quot;</span>
    <span class="k">pass</span>


<span class="k">class</span> <span class="nc">DataMismatch</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Raised when dimensions of sample/ID names do not match dimension of</span>
<span class="sd">     corresponding data&quot;&quot;&quot;</span>
    <span class="k">pass</span>


<div class="viewcode-block" id="ReadData"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData">[docs]</a><span class="k">class</span> <span class="nc">ReadData</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate object containing all datasets relevant for the analysis.</span>
<span class="sd">    For variance decomposition, at least phenotypes and relatedness estimates</span>
<span class="sd">    need to be specified.</span>
<span class="sd">    For association testing with LMM, at least phenotype, relatedness estimates</span>
<span class="sd">    and genotypes need to be read.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pheno_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_ID</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">covs_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">relatedness_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pc_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">snps</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">geno_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">position</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="ReadData.getPhenotypes"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData.getPhenotypes">[docs]</a>    <span class="k">def</span> <span class="nf">getPhenotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_pheno</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delim</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reads [`N` x `P`] phenotype file; file ending must be either .txt or </span>
<span class="sd">        .csv</span>

<span class="sd">        Arguments:</span>
<span class="sd">            file_pheno (string):</span>
<span class="sd">                path to [(`N`+1) x (`P`+1)] phenotype file with: [`N`] sample </span>
<span class="sd">                IDs in the first column and [`P`] phenotype IDs in the first row</span>
<span class="sd">            delim (string): </span>
<span class="sd">                delimiter of phenotype file, one of &quot; &quot;, &quot;,&quot;, &quot;\t&quot;</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the ReadData instance:</span>

<span class="sd">                - **self.phenotypes** (np.array):</span>
<span class="sd">                  [`N` x `P`] phenotype matrix</span>

<span class="sd">        Examples:</span>

<span class="sd">            .. doctest::</span>

<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.reader import ReadData</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.utils import file_type</span>
<span class="sd">                &gt;&gt;&gt; data = ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_pheno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                                &#39;io/test/data/pheno.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getPhenotypes(file_pheno=file_pheno)</span>
<span class="sd">                &gt;&gt;&gt; data.phenotypes.index[:3]</span>
<span class="sd">                Index([u&#39;ID_1&#39;, u&#39;ID_2&#39;, u&#39;ID_3&#39;], dtype=&#39;object&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.phenotypes.columns[:3]</span>
<span class="sd">                Index([u&#39;trait_1&#39;, u&#39;trait_2&#39;, u&#39;trait_3&#39;], dtype=&#39;object&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.phenotypes.values[:3,:3]</span>
<span class="sd">                array([[-1.56760036, -1.5324513 ,  1.17789321],</span>
<span class="sd">                       [-0.85655034,  0.48358151,  1.35664966],</span>
<span class="sd">                       [ 0.10772832, -0.02262884, -0.27963328]])</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">file_pheno</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">MissingInput</span><span class="p">(</span><span class="s1">&#39;No phenotype file specified&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file_type</span><span class="p">(</span><span class="n">file_pheno</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="s2">&quot;delim&quot;</span><span class="p">:</span> 
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Supplied phenotype file is neither .csv &#39;</span> 
                        <span class="s1">&#39;nor .txt&#39;</span><span class="p">)</span>
            <span class="n">verboseprint</span><span class="p">(</span>
                <span class="s2">&quot;Reading phenotypes from </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">file_pheno</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                    <span class="n">file_pheno</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">delim</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> could not be opened&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_pheno</span><span class="p">))</span></div>

<div class="viewcode-block" id="ReadData.getCovariates"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData.getCovariates">[docs]</a>    <span class="k">def</span> <span class="nf">getCovariates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_covariates</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delim</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reads [`N` x `K`] covariate matrix with [`N`] samples and [`K`] </span>
<span class="sd">        covariates.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            file_covariates (string):</span>
<span class="sd">                [`N` x (`K` +1)] covariates file with [`N`] sample IDs in the </span>
<span class="sd">                first column</span>
<span class="sd">            delim (string): </span>
<span class="sd">                delimiter of covariates file, one of &quot; &quot;, &quot;,&quot;, &quot;\t&quot;</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the ReadData instance:</span>

<span class="sd">                - **self.covariates** (np.array):</span>
<span class="sd">                  [`N` x `K`] covariates matrix</span>

<span class="sd">        Examples:</span>

<span class="sd">            .. doctest::</span>

<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.reader import ReadData</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.utils import file_type</span>
<span class="sd">                &gt;&gt;&gt; data = ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_covs = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                               &#39;io/test/data/covs.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getCovariates(file_covariates=file_covs)</span>
<span class="sd">                &gt;&gt;&gt; data.covariates.index[:3]</span>
<span class="sd">                Index([u&#39;ID_1&#39;, u&#39;ID_2&#39;, u&#39;ID_3&#39;], dtype=&#39;object&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.covariates.values[:3,:3]</span>
<span class="sd">                array([[ 0.92734699,  1.59767659, -0.67263682],</span>
<span class="sd">                       [ 0.57061985, -0.84679736, -1.11037123],</span>
<span class="sd">                       [ 0.44201204, -1.61499228,  0.23302345]])</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">file_covariates</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file_type</span><span class="p">(</span><span class="n">file_covariates</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="s1">&#39;delim&#39;</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Supplied covariate file is not .csv or .txt&#39;</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_covariates</span><span class="p">,</span> 
                        <span class="n">sep</span><span class="o">=</span><span class="n">delim</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Reading covariates file&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> could not be opened&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_covariates</span><span class="p">))</span>
            <span class="c1"># append column of 1&#39;s to adjust for mean of covariates</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">sp</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">covariates</span><span class="o">.</span><span class="n">index</span><span class="p">)],</span>
                 <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;No covariates set&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">covariates</span> <span class="o">=</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="ReadData.getRelatedness"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData.getRelatedness">[docs]</a>    <span class="k">def</span> <span class="nf">getRelatedness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_relatedness</span><span class="p">,</span> <span class="n">delim</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read file of [`N` x `N`] pairwise relatedness estimates of [`N`] </span>
<span class="sd">        samples.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            file_relatedness (string):</span>
<span class="sd">                [(`N` + `1`) x N] .csv file with: [`N`] sample IDs in the first</span>
<span class="sd">                row</span>
<span class="sd">            delim (string): </span>
<span class="sd">                delimiter of covariate file, one of &quot; &quot;, &quot;,&quot;, &quot;\t&quot;</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the ReadData instance:</span>

<span class="sd">                - **self.relatedness** (np.array):</span>
<span class="sd">                  [`N` x `N`] relatedness matrix</span>
<span class="sd">        Examples:</span>

<span class="sd">            .. doctest::</span>

<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.reader import ReadData</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.utils import file_type</span>
<span class="sd">                &gt;&gt;&gt; data = ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_relatedness = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/relatedness.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getRelatedness(file_relatedness=file_relatedness)</span>
<span class="sd">                &gt;&gt;&gt; data.relatedness.index[:3]</span>
<span class="sd">                Index([u&#39;ID_1&#39;, u&#39;ID_2&#39;, u&#39;ID_3&#39;], dtype=&#39;object&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.relatedness.columns[:3]</span>
<span class="sd">                Index([u&#39;ID_1&#39;, u&#39;ID_2&#39;, u&#39;ID_3&#39;], dtype=&#39;object&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.relatedness.values[:3,:3]</span>
<span class="sd">                array([[1.00892922e+00, 2.00758504e-04, 4.30499103e-03],</span>
<span class="sd">                       [2.00758504e-04, 9.98944885e-01, 4.86487318e-03],</span>
<span class="sd">                       [4.30499103e-03, 4.86487318e-03, 9.85787665e-01]])</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">file_relatedness</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">MissingInput</span><span class="p">(</span><span class="s1">&#39;No relatedness data specified&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">file_type</span><span class="p">(</span><span class="n">file_relatedness</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="s1">&#39;delim&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Supplied relatedness file is not .csv or .txt&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_relatedness</span><span class="p">,</span> 
                    <span class="n">sep</span><span class="o">=</span><span class="n">delim</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">relatedness</span><span class="o">.</span><span class="n">columns</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> could not be opened&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_relatedness</span><span class="p">))</span>
        <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Reading relationship matrix&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span></div>

<div class="viewcode-block" id="ReadData.getPCs"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData.getPCs">[docs]</a>    <span class="k">def</span> <span class="nf">getPCs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_pcs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">nrpcs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delim</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reads file with [`N` x `PC`] matrix of [`PC`] principal components from </span>
<span class="sd">        the genotypes of [`N`] samples.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            file_pcs (string):</span>
<span class="sd">                [`N` x (`PC` +1)] PCA file with [`N`] sample IDs in the first</span>
<span class="sd">                column</span>
<span class="sd">            delim (string): </span>
<span class="sd">                delimiter of PCA file, one of &quot; &quot;, &quot;,&quot;, &quot;\t&quot;</span>
<span class="sd">            nrpcs (integer):</span>
<span class="sd">                Number of PCs to use (uses first nrpcs  principal components)</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the ReadData instance:</span>

<span class="sd">                - **self.pcs** (np.array):</span>
<span class="sd">                  [`N` x `PC`] principal component matrix</span>

<span class="sd">        Examples:</span>

<span class="sd">            .. doctest::</span>

<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.reader import ReadData</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.utils import file_type</span>
<span class="sd">                &gt;&gt;&gt; data = ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; file_pcs = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/pcs.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getPCs(file_pcs=file_pcs, nrpcs=10, delim=&quot; &quot;)</span>
<span class="sd">                &gt;&gt;&gt; data.pcs.index[:3]</span>
<span class="sd">                Index([u&#39;ID_1&#39;, u&#39;ID_2&#39;, u&#39;ID_3&#39;], dtype=&#39;object&#39;, name=0)</span>
<span class="sd">                &gt;&gt;&gt; data.pcs.values[:3,:3]</span>
<span class="sd">                array([[-0.02632738, -0.05791269, -0.03619099],</span>
<span class="sd">                       [ 0.00761785,  0.00538956,  0.00624196],</span>
<span class="sd">                       [ 0.01069307, -0.0205066 ,  0.02299996]])</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">file_type</span><span class="p">(</span><span class="n">file_pcs</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="s1">&#39;delim&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">FormatError</span><span class="p">(</span><span class="s1">&#39;Supplied PCA file is not .csv or .txt&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">file_pcs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Reading PCs&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_pcs</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> 
                            <span class="n">sep</span><span class="o">=</span><span class="n">delim</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">nrpcs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">verboseprint</span><span class="p">(</span>
                    <span class="s2">&quot;Extracting first </span><span class="si">%s</span><span class="s2"> pcs&quot;</span> <span class="o">%</span> <span class="n">nrpcs</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="n">nrpcs</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;No pcs set&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="ReadData.getGenotypes"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData.getGenotypes">[docs]</a>    <span class="k">def</span> <span class="nf">getGenotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_genotypes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delim</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reads genotype file in the following formats: plink (.bed, .bim, .fam),  </span>
<span class="sd">        gen (.gen, .sample) or comma-separated values (.csv) file.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            file_geno (string):</span>
<span class="sd">                path to phenotype file in .plink or .csv format</span>
<span class="sd">                - **plink format**:</span>
<span class="sd">                  </span>
<span class="sd">                  as specified in the plink `user manual &lt;https://www.cog-genomics.org/plink/1.9/input&gt;`_, </span>
<span class="sd">                  binary plink format with .bed, .fam and .bim file</span>

<span class="sd">                - **.csv format**:</span>

<span class="sd">                  - [(`NrSNP` + 1) x (`N`+1)] .csv file with: [`N`] sample IDs</span>
<span class="sd">                    in the first row and [`NrSNP`] genotype IDs in the first</span>
<span class="sd">                    column</span>
<span class="sd">                  - sample IDs should be of type: chrom-pos-rsID for instance</span>
<span class="sd">                    22-50714616-rs6010226</span>
<span class="sd">            delim (string): </span>
<span class="sd">                delimiter of genotype file (when text format), one of &quot; &quot;, &quot;,&quot;, </span>
<span class="sd">                &quot;\t&quot;</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the ReadData instance:</span>

<span class="sd">                - **self.genotypes** (np.array):</span>
<span class="sd">                  [`N` x `NrSNPs`] genotype matrix</span>
<span class="sd">                - **self.genotypes_info** (pd.dataframe):</span>
<span class="sd">                  [`NrSNPs` x 2] dataframe with columns &#39;chrom&#39; and &#39;pos&#39;, and</span>
<span class="sd">                  rsIDs as index</span>

<span class="sd">        Examples:</span>

<span class="sd">            .. doctest::</span>

<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import reader</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.utils import file_type</span>
<span class="sd">                &gt;&gt;&gt; data = reader.ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; # Read genotypes in delim-format</span>
<span class="sd">                &gt;&gt;&gt; file_geno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...     &#39;io/test/data/genotypes.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getGenotypes(file_genotypes=file_geno)</span>
<span class="sd">                &gt;&gt;&gt; data.genotypes.index[:4]</span>
<span class="sd">                Index([u&#39;ID_1&#39;, u&#39;ID_2&#39;, u&#39;ID_3&#39;, u&#39;ID_4&#39;], dtype=&#39;object&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.genotypes.shape</span>
<span class="sd">                (1000, 20)</span>
<span class="sd">                &gt;&gt;&gt; data.genotypes.values[:5,:5]</span>
<span class="sd">                array([[0., 0., 0., 0., 0.],</span>
<span class="sd">                       [0., 0., 0., 0., 0.],</span>
<span class="sd">                       [0., 0., 0., 0., 0.],</span>
<span class="sd">                       [2., 1., 0., 0., 0.],</span>
<span class="sd">                       [1., 0., 0., 0., 0.]])</span>
<span class="sd">                &gt;&gt;&gt; data.genotypes_info[:5]</span>
<span class="sd">                           chrom       pos</span>
<span class="sd">                rs1601111      3  88905003</span>
<span class="sd">                rs13270638     8  20286021</span>
<span class="sd">                rs75132935     8  76564608</span>
<span class="sd">                rs72668606     8  79733124</span>
<span class="sd">                rs55770986     7   2087823</span>
<span class="sd">                &gt;&gt;&gt; ### read genotypes in plink format</span>
<span class="sd">                &gt;&gt;&gt; file_geno = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...     &#39;io/test/data/genotypes&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getGenotypes(file_genotypes=file_geno)</span>
<span class="sd">                &gt;&gt;&gt; data.genotypes.values[:5,:5]</span>
<span class="sd">                array([[0., 0., 0., 0., 0.],</span>
<span class="sd">                       [0., 0., 0., 0., 0.],</span>
<span class="sd">                       [0., 0., 0., 0., 0.],</span>
<span class="sd">                       [2., 1., 0., 0., 0.],</span>
<span class="sd">                       [1., 0., 0., 0., 0.]])</span>
<span class="sd">                &gt;&gt;&gt; data.genotypes_info[:5]</span>
<span class="sd">                           chrom       pos</span>
<span class="sd">                rs1601111      3  88905003</span>
<span class="sd">                rs13270638     8  20286021</span>
<span class="sd">                rs75132935     8  76564608</span>
<span class="sd">                rs72668606     8  79733124</span>
<span class="sd">                rs55770986     7   2087823</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">file_genotypes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">MissingInput</span><span class="p">(</span><span class="s1">&#39;No genotypes data specified&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">file_type</span><span class="p">(</span><span class="n">file_genotypes</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;delim&#39;</span><span class="p">,</span> <span class="s1">&#39;bed&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">FormatError</span><span class="p">((</span><span class="s1">&#39;Supplied genotype file is neither in plink nor &#39;</span>
                    <span class="s1">&#39;.csv/.txt format&#39;</span><span class="p">))</span>
        <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Reading genotypes from </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">file_genotypes</span><span class="p">,</span> 
                <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
        
        
        <span class="k">if</span> <span class="n">file_type</span><span class="p">(</span><span class="n">file_genotypes</span><span class="p">)</span> <span class="ow">is</span> <span class="s1">&#39;delim&#39;</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">genotypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                    <span class="n">file_genotypes</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">delim</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> could not be opened&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_genotypes</span><span class="p">))</span>

            <span class="n">info</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotypes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

            <span class="n">genotypes_info</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">snp_ID</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
                <span class="n">split</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">info</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">))</span>
                <span class="n">snp_ID</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">split</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                <span class="n">genotypes_info</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">split</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]])</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">=</span> <span class="n">genotypes</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genotypes_info</span><span class="p">),</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;pos&#39;</span><span class="p">],</span>
                <span class="n">index</span><span class="o">=</span><span class="n">snp_ID</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">file_type</span><span class="p">(</span><span class="n">file_genotypes</span><span class="p">)</span> <span class="ow">is</span> <span class="s1">&#39;bed&#39;</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="p">(</span><span class="n">bim</span><span class="p">,</span> <span class="n">fam</span><span class="p">,</span> <span class="n">bed</span><span class="p">)</span> <span class="o">=</span> <span class="n">read_plink</span><span class="p">(</span><span class="n">file_genotypes</span><span class="p">,</span>
                        <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> could not be opened&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_genotypes</span><span class="p">))</span>
            
            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">compute</span><span class="p">())</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">fam</span><span class="o">.</span><span class="n">iid</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">bim</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">bim</span><span class="o">.</span><span class="n">pos</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;pos&#39;</span><span class="p">],</span>
                <span class="n">index</span><span class="o">=</span><span class="n">bim</span><span class="o">.</span><span class="n">snp</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotypes_info</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="ReadData.getVarianceComponents"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData.getVarianceComponents">[docs]</a>    <span class="k">def</span> <span class="nf">getVarianceComponents</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_Cg</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">file_Cn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delim_cg</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
            <span class="n">delim_cn</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reads a comma-separated files with [`P` x `P`] matrices of [`P`] trait</span>
<span class="sd">        covariance estimates.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            file_Cg (string):</span>
<span class="sd">                [`P` x `P`] .csv file with [`P`] trait covariance estimates of</span>
<span class="sd">                the genetic component</span>
<span class="sd">            file_Cn (string):</span>
<span class="sd">                [`P` x `P`] .csv file with [`P`] trait covariance estimates of</span>
<span class="sd">                the non-genetic (noise) component</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>
<span class="sd">                updated the following attributes of the ReadData instance:</span>

<span class="sd">                - **self.Cg** (np.array):</span>
<span class="sd">                  [`P` x `P`] matrix with trait covariance of the genetic</span>
<span class="sd">                  component</span>
<span class="sd">                - **self.Cn** (np.array):</span>
<span class="sd">                  [`P` x `P`] matrix with trait covariance of the non-genetic</span>
<span class="sd">                  (noise) component</span>

<span class="sd">        Examples:</span>

<span class="sd">            .. doctest::</span>

<span class="sd">                &gt;&gt;&gt; from pkg_resources import resource_filename</span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io.reader import ReadData</span>
<span class="sd">                &gt;&gt;&gt; data = ReadData()</span>
<span class="sd">                &gt;&gt;&gt; file_Cg = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/Cg.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; file_Cn = resource_filename(&#39;limmbo&#39;,</span>
<span class="sd">                ...                     &#39;io/test/data/Cn.csv&#39;)</span>
<span class="sd">                &gt;&gt;&gt; data.getVarianceComponents(file_Cg=file_Cg, file_Cn=file_Cn)</span>
<span class="sd">                &gt;&gt;&gt; data.Cg.shape</span>
<span class="sd">                (10, 10)</span>
<span class="sd">                &gt;&gt;&gt; data.Cn.shape</span>
<span class="sd">                (10, 10)</span>
<span class="sd">                &gt;&gt;&gt; data.Cg[:3,:3]</span>
<span class="sd">                array([[ 0.45446454, -0.21084613,  0.01440468],</span>
<span class="sd">                       [-0.21084613,  0.11443656,  0.01250233],</span>
<span class="sd">                       [ 0.01440468,  0.01250233,  0.02347906]])</span>
<span class="sd">                &gt;&gt;&gt; data.Cn[:3,:3]</span>
<span class="sd">                array([[ 0.53654803, -0.14392748, -0.45483001],</span>
<span class="sd">                       [-0.14392748,  0.88793093,  0.30539822],</span>
<span class="sd">                       [-0.45483001,  0.30539822,  0.97785614]])</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">file_Cg</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">file_Cn</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span>
                <span class="p">(</span><span class="s2">&quot;No variance components supplied, estimate variance components &quot;</span>
                 <span class="s2">&quot;before lmm test&quot;</span><span class="p">),</span>
                <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">file_Cg</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">file_Cn</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;Both variant components need to be supplied:&#39;</span><span class="p">,</span>
                          <span class="s1">&#39;Cg is </span><span class="si">%s</span><span class="s1"> and Cn is </span><span class="si">%s</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">%</span> <span class="p">(</span><span class="n">file_Cg</span><span class="p">,</span> <span class="n">file_Cn</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">Cg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="n">pd</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_Cg</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">delim_cg</span><span class="p">))</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> could not be opened&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_Cg</span><span class="p">))</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">Cn</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="n">pd</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_Cn</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">delim_cn</span><span class="p">))</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> could not be opened&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_Cn</span><span class="p">))</span></div>
    
<div class="viewcode-block" id="ReadData.getTraitSubset"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData.getTraitSubset">[docs]</a>    <span class="k">def</span> <span class="nf">getTraitSubset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">traitstring</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Limit analysis to specific subset of traits</span>
<span class="sd">        </span>
<span class="sd">        Arguments:</span>
<span class="sd">            traitstring (string):</span>
<span class="sd">                comma-separated trait numbers (for single traits) or hyphen-</span>
<span class="sd">                separated trait numbers (for trait ranges) or combination of</span>
<span class="sd">                both for trait selection (1-based)</span>
<span class="sd">        </span>
<span class="sd">        Returns:</span>
<span class="sd">            (numpy array)</span>
<span class="sd">                array containing list of trait IDs</span>
<span class="sd">        </span>
<span class="sd">        Examples:</span>
<span class="sd">        </span>
<span class="sd">            .. doctest::</span>
<span class="sd">        </span>
<span class="sd">                &gt;&gt;&gt; from limmbo.io import reader</span>
<span class="sd">                &gt;&gt;&gt; data = reader.ReadData(verbose=False)</span>
<span class="sd">                &gt;&gt;&gt; traitlist = data.getTraitSubset(&quot;1,3,5,7-10&quot;)</span>
<span class="sd">                &gt;&gt;&gt; print traitlist</span>
<span class="sd">                [0 2 4 6 7 8 9]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">traitstring</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s1">&#39;No trait subset chosen&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">verboseprint</span><span class="p">(</span><span class="s1">&#39;Chose subset of </span><span class="si">{}</span><span class="s1"> traits&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">traitstring</span><span class="p">),</span> 
                    <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="n">search</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;[^0-9,-]&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">search</span>
            <span class="k">if</span> <span class="nb">bool</span><span class="p">(</span><span class="n">search</span><span class="p">(</span><span class="n">traitstring</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="n">FormatError</span><span class="p">((</span><span class="s1">&#39;Traitstring can only contain integers &#39;</span>
                        <span class="s1">&#39;(0-9), comma (,) and hyphen (-), but </span><span class="si">{}</span><span class="s1">&#39;</span>
                        <span class="s1">&#39;provided&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">traitstring</span><span class="p">))</span>
            <span class="n">traitslist</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">traitstring</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span> <span class="p">]</span>
            <span class="n">traitsarray</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">traitslist</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">traitsarray</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="p">[</span><span class="n">traitsarray</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span> <span class="p">]</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">traitsarray</span><span class="p">)</span></div>

<div class="viewcode-block" id="ReadData.getSampleSubset"><a class="viewcode-back" href="../../../input.html#limmbo.io.reader.ReadData.getSampleSubset">[docs]</a>    <span class="k">def</span> <span class="nf">getSampleSubset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_samplelist</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">samplelist</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read file or string with subset of sample IDs to analyse.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            file_samplelist (string):</span>
<span class="sd">                &quot;path/to/file_samplelist&quot;: file contains subset sample IDs with</span>
<span class="sd">                one ID per line, no header.</span>
<span class="sd">            samplestring (string): </span>
<span class="sd">                comma-separated list of sample IDs e.g. &quot;ID1,ID2,ID5,ID10&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (numpy array)</span>
<span class="sd">                array containing list of sample IDs</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">file_samplelist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">samplelist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
           <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;Only one of file_samplelist or samplelist can &quot;</span>
                    <span class="s2">&quot;be specified&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">file_samplelist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">samplelist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file_samplelist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Read sample list from file&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">samplelist</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parsers</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                        <span class="n">file_samplelist</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> could not be opened&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">file_samplelist</span><span class="p">))</span>
                <span class="n">samplelist</span> <span class="o">=</span> <span class="n">samplelist</span><span class="o">.</span><span class="n">index</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Split sample string&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
                <span class="n">samplelist</span> <span class="o">=</span> <span class="n">samplelist</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>

            <span class="n">verboseprint</span><span class="p">(</span><span class="s2">&quot;Number of samples in sample list: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">samplelist</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">samplelist</span></div></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Hannah V. Meyer.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'b',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>