

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Association analysis &mdash; LiMMBo b documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="LiMMBo b documentation" href="index.html"/>
        <link rel="next" title="Command-line interface" href="cmd.html"/>
        <link rel="prev" title="Variance decomposition" href="vd.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> LiMMBo
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Data input</a></li>
<li class="toctree-l1"><a class="reference internal" href="vd.html">Variance decomposition</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Association analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="cmd.html">Command-line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">Additional functions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">LiMMBo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Association analysis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/gwas.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-limmbo.core.gwas">
<span id="association-analysis"></span><h1>Association analysis<a class="headerlink" href="#module-limmbo.core.gwas" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="limmbo.core.gwas.GWAS">
<em class="property">class </em><code class="descclassname">limmbo.core.gwas.</code><code class="descname">GWAS</code><span class="sig-paren">(</span><em>datainput</em>, <em>seed=10</em>, <em>verbose=True</em>, <em>searchDelta=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/limmbo/core/gwas.html#GWAS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#limmbo.core.gwas.GWAS" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="limmbo.core.gwas.GWAS.computeEmpiricalP">
<code class="descname">computeEmpiricalP</code><span class="sig-paren">(</span><em>pvalues</em>, <em>nrpermutations=1000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/limmbo/core/gwas.html#GWAS.computeEmpiricalP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#limmbo.core.gwas.GWAS.computeEmpiricalP" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute empirical p-values: permute the genotypes, do the
association test, record if permuted p-value of SNP is smaller than
original p-value. Sum these occurrences and divide by total number of
permutation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pvalues</strong> (<em>array-like</em>) – [<cite>P</cite> x <cite>NrSNP</cite>] (single-trait) or [1 x <cite>NrSNP</cite>] (multi-trait)
array of p-values</li>
<li><strong>nrpermutations</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) – number of permutations; 1/nrpermutations is the maximum level
of significance (alpha)to test for, e.g.
nrpermuations=100 -&gt; alpha=0.01</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">[<cite>P</cite> x <cite>NrSNP</cite>] (single-trait) or [1 x <cite>NrSNP</cite>] (multi-trait)
array of empirical p-values</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">(numpy array)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="limmbo.core.gwas.GWAS.computeFDR">
<code class="descname">computeFDR</code><span class="sig-paren">(</span><em>fdr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/limmbo/core/gwas.html#GWAS.computeFDR"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#limmbo.core.gwas.GWAS.computeFDR" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an empirical p-values distribution by permuting the genotypes
and running the association test on these (10 random permuations).
Recored the observed pvalues, order by rank and find the rank of the
desired FDR to determine the empirical FDR.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fdr</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) – desired fdr threshold</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">tuple containing:<blockquote>
<div><ul class="simple">
<li><strong>empirical fdr</strong> (float):</li>
<li><strong>empirical_pvalue_distribution</strong> (numpy array):
array of empirical p-values for 10 permutations tests</li>
</ul>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">(<a class="reference external" href="https://docs.python.org/2/library/functions.html#tuple" title="(in Python v2.7)">tuple</a>)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="limmbo.core.gwas.GWAS.manhattanQQ">
<code class="descname">manhattanQQ</code><span class="sig-paren">(</span><em>results</em>, <em>colourS='DarkBLue'</em>, <em>colourNS='Orange'</em>, <em>alphaS=0.5</em>, <em>alphaNS=0.1</em>, <em>thr_plotting=None</em>, <em>saveTo=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/limmbo/core/gwas.html#GWAS.manhattanQQ"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#limmbo.core.gwas.GWAS.manhattanQQ" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot manhattan and quantile-quantile plot of association results.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>results</strong> (<em>dictionary</em>) – dictionary generated via runAssociation analysis</li>
<li><strong>colourS</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a><em>, </em><em>optional</em>) – colour of significant points</li>
<li><strong>colourNS</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a><em>, </em><em>optional</em>) – colour of non-significant points</li>
<li><strong>alphaS</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – plotting transparency of significant points</li>
<li><strong>alphaNS</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – plotting transparency of non-significant points</li>
<li><strong>thr_plotting</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a><em>, </em><em>optional</em>) – y-intercept for horizontal line as a marker for significance</li>
<li><strong>saveTo</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a><em>, </em><em>optional</em>) – /path/to/output/directory to automatically save plot as pdf;
needs user writing permission.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">(None)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="limmbo.core.gwas.GWAS.runAssociationAnalysis">
<code class="descname">runAssociationAnalysis</code><span class="sig-paren">(</span><em>mode</em>, <em>setup='lmm'</em>, <em>adjustSingleTrait=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/limmbo/core/gwas.html#GWAS.runAssociationAnalysis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#limmbo.core.gwas.GWAS.runAssociationAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysing the association between phenotypes, genotypes, optional
covariates and random genetic effects.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – pecifies the type of linear model: either ‘multitrait’ for
multivariate analysis or ‘singletrait’ for univariate analysis.</li>
<li><strong>setup</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – specifies the linear model setup: either ‘lmm’ for linear mixed
model or ‘lm’ for a simple linear model.</li>
<li><strong>adjustSingleTrait</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – Method to adjust single-trait association p-values for testing
multiple traits; If None (default) no adjusting. Options are
‘bonferroni’ (for bonferroni correction’) or ‘effective’ (for
correcting for the effective number of tests as described in
<a class="reference external" href="http://onlinelibrary.wiley.com/doi/10.1002/gepi.20408/abstract">(Galwey,2009)</a>.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><p>dictionary containing:</p>
<ul class="simple">
<li><strong>lm</strong> (<code class="xref py py-class docutils literal"><span class="pre">limix.qtl.LMM</span></code>):
LIMIX LMM object</li>
<li><strong>pvalues</strong> (numpy array):
[<cite>NrSNP</cite> x <cite>P</cite>] (when mode is singletrait) or [1 x`NrSNP`]
array of p-values.</li>
<li><strong>betas</strong> (numpy array):
[<cite>NrSNP</cite> x <cite>P</cite>] array of effect size estimates per SNP across
all traits.</li>
<li><strong>pvalues_adjust</strong> (numpy array):
only returned if mode is ‘singletrait’ and ‘adjustSingleTrait’
is not None; contains single-trait p-values adjusted for the
number of single-trait analyses conducted.</li>
</ul>
</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">(dictionary)</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pkg_resources</span> <span class="k">import</span> <span class="n">resource_filename</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">limmbo.io.reader</span> <span class="k">import</span> <span class="n">ReadData</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">limmbo.io.input</span> <span class="k">import</span> <span class="n">InputData</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">limmbo.core.gwas</span> <span class="k">import</span> <span class="n">GWAS</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">ReadData</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_pheno</span> <span class="o">=</span> <span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;limmbo&#39;</span><span class="p">,</span>
<span class="gp">... </span>                               <span class="s1">&#39;io/test/data/pheno.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_geno</span> <span class="o">=</span> <span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;limmbo&#39;</span><span class="p">,</span>
<span class="gp">... </span>                               <span class="s1">&#39;io/test/data/genotypes.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_relatedness</span> <span class="o">=</span> <span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;limmbo&#39;</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="s1">&#39;io/test/data/relatedness.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_covs</span> <span class="o">=</span> <span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;limmbo&#39;</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="s1">&#39;io/test/data/covs.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_Cg</span> <span class="o">=</span> <span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;limmbo&#39;</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="s1">&#39;io/test/data/Cg.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_Cn</span> <span class="o">=</span> <span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;limmbo&#39;</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="s1">&#39;io/test/data/Cn.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">getPhenotypes</span><span class="p">(</span><span class="n">file_pheno</span><span class="o">=</span><span class="n">file_pheno</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">getCovariates</span><span class="p">(</span><span class="n">file_covariates</span><span class="o">=</span><span class="n">file_covs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">getRelatedness</span><span class="p">(</span><span class="n">file_relatedness</span><span class="o">=</span><span class="n">file_relatedness</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">getGenotypes</span><span class="p">(</span><span class="n">file_genotypes</span><span class="o">=</span><span class="n">file_geno</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">getVarianceComponents</span><span class="p">(</span><span class="n">file_Cg</span><span class="o">=</span><span class="n">file_Cg</span><span class="p">,</span>
<span class="gp">... </span>                           <span class="n">file_Cn</span><span class="o">=</span><span class="n">file_Cn</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span> <span class="o">=</span> <span class="n">InputData</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span><span class="o">.</span><span class="n">addPhenotypes</span><span class="p">(</span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">phenotypes</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span><span class="o">.</span><span class="n">addRelatedness</span><span class="p">(</span><span class="n">relatedness</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">relatedness</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span><span class="o">.</span><span class="n">addCovariates</span><span class="p">(</span><span class="n">covariates</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">covariates</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span><span class="o">.</span><span class="n">addGenotypes</span><span class="p">(</span><span class="n">genotypes</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">genotypes</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="n">genotypes_info</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">genotypes_info</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span><span class="o">.</span><span class="n">addVarianceComponents</span><span class="p">(</span><span class="n">Cg</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">Cg</span><span class="p">,</span> <span class="n">Cn</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">Cn</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span><span class="o">.</span><span class="n">commonSamples</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span><span class="o">.</span><span class="n">regress</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indata</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">transform</span><span class="o">=</span><span class="s2">&quot;scale&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gwas</span> <span class="o">=</span> <span class="n">GWAS</span><span class="p">(</span><span class="n">datainput</span><span class="o">=</span><span class="n">indata</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example of multi-trait single-variant association testing</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># using a linear mixed model.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resultsAssociation</span> <span class="o">=</span> <span class="n">gwas</span><span class="o">.</span><span class="n">runAssociationAnalysis</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">setup</span><span class="o">=</span><span class="s2">&quot;lmm&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;multitrait&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resultsAssociation</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[&#39;lm&#39;, &#39;betas&#39;, &#39;pvalues&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resultsAssociation</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(1, 20)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;</span><span class="si">{:0.3e}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">resultsAssociation</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
<span class="go">&#39;4.584e-09&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resultsAssociation</span><span class="p">[</span><span class="s1">&#39;betas&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(10, 20)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example of single-trait single-variant association testing</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># using a linear mixed model.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resultsAssociation</span> <span class="o">=</span> <span class="n">gwas</span><span class="o">.</span><span class="n">runAssociationAnalysis</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">setup</span><span class="o">=</span><span class="s2">&quot;lmm&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;singletrait&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">adjustSingleTrait</span> <span class="o">=</span> <span class="s2">&quot;effective&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resultsAssociation</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[&#39;pvalues_adjust&#39;, &#39;lm&#39;, &#39;betas&#39;, &#39;pvalues&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resultsAssociation</span><span class="p">[</span><span class="s1">&#39;pvalues&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(10, 20)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resultsAssociation</span><span class="p">[</span><span class="s1">&#39;betas&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(10, 20)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;</span><span class="si">{:0.3e}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">resultsAssociation</span><span class="p">[</span><span class="s1">&#39;pvalues_adjust&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
<span class="go">&#39;2.262e-03&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="limmbo.core.gwas.GWAS.saveAssociationResults">
<code class="descname">saveAssociationResults</code><span class="sig-paren">(</span><em>results</em>, <em>outdir</em>, <em>name=''</em>, <em>pvalues_empirical=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/limmbo/core/gwas.html#GWAS.saveAssociationResults"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#limmbo.core.gwas.GWAS.saveAssociationResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves results of association analyses.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>results</strong> (<em>dictionary</em>) – <p>dictionary generated via runAssociation analysis, containing:</p>
<ul>
<li><strong>lm</strong> (<code class="xref py py-class docutils literal"><span class="pre">limix.qtl.LMM</span></code>):
LIMIX LMM object</li>
<li><strong>pvalues</strong> (numpy array):
[<cite>P</cite> x <cite>NrSNP</cite>] array of p-values</li>
<li><strong>betas</strong> (numpy array):
[<cite>P</cite> x <cite>NrSNP</cite>] array of effect size estimates per SNP
across all traits</li>
<li><strong>pvalues_adjust</strong> (numpy array):
only returned mode == singletrait and if ‘adjustSingleTrait’
is not None; contains single-trait p-values adjusted for the
number of single-trait analyses conducted</li>
</ul>
</li>
<li><strong>outdir</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – ‘/path/to/output/directory’; needs user writing permission</li>
<li><strong>name</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – input name specific name, such as chromosome used in analysis</li>
<li><strong>pvalues_empirical</strong> (<em>pandas dataframe</em><em>, </em><em>optional</em>) – empirical pvalues via adjustSingleTrait</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">(None)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="cmd.html" class="btn btn-neutral float-right" title="Command-line interface" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="vd.html" class="btn btn-neutral" title="Variance decomposition" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Hannah V. Meyer.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'b',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>